{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "f5076ec1-f123-46b3-a92c-eec1f3acf6fe",
   "metadata": {},
   "source": [
    "# Benchmarking tokenizers for cross-tissue generalization"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "20f94504-650b-470b-bb51-32c712d9f7aa",
   "metadata": {},
   "source": [
    "Here, we demonstrate how to use <span style=\"font-variant:small-caps; font-size: 16px\">Heimdall</span> for benchmarking the impact of tokenizer choice on cell-type annotation performance in a challenging cross-tissue evaluation setting. For this evaluation, we use a subset of the [**scTab** dataset](https://www.nature.com/articles/s41467-024-51059-5)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "53af211e-cacc-4b91-ab85-57a6703aece1",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-11-18T11:12:10.116496Z",
     "iopub.status.busy": "2025-11-18T11:12:10.116298Z",
     "iopub.status.idle": "2025-11-18T11:12:10.119558Z",
     "shell.execute_reply": "2025-11-18T11:12:10.119015Z",
     "shell.execute_reply.started": "2025-11-18T11:12:10.116481Z"
    }
   },
   "outputs": [],
   "source": [
    "import hydra\n",
    "import Heimdall"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "f0460b09-04fa-4b5b-83fe-3a77f3190a1d",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-11-18T11:13:38.488666Z",
     "iopub.status.busy": "2025-11-18T11:13:38.488473Z",
     "iopub.status.idle": "2025-11-18T11:13:38.490907Z",
     "shell.execute_reply": "2025-11-18T11:13:38.490374Z",
     "shell.execute_reply.started": "2025-11-18T11:13:38.488651Z"
    }
   },
   "outputs": [],
   "source": [
    "from matplotlib import pyplot as plt\n",
    "import matplotlib\n",
    "import seaborn as sns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "97a8d119-d4a9-4e7d-8d8f-96e5de18b30f",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-11-18T11:13:38.719345Z",
     "iopub.status.busy": "2025-11-18T11:13:38.719191Z",
     "iopub.status.idle": "2025-11-18T11:13:38.721294Z",
     "shell.execute_reply": "2025-11-18T11:13:38.720811Z",
     "shell.execute_reply.started": "2025-11-18T11:13:38.719332Z"
    }
   },
   "outputs": [],
   "source": [
    "import scanpy as sc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "3de731a8-f86a-48f1-9a9e-f602c2446607",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-11-18T11:13:39.125504Z",
     "iopub.status.busy": "2025-11-18T11:13:39.125364Z",
     "iopub.status.idle": "2025-11-18T11:13:39.128348Z",
     "shell.execute_reply": "2025-11-18T11:13:39.127846Z",
     "shell.execute_reply.started": "2025-11-18T11:13:39.125492Z"
    }
   },
   "outputs": [],
   "source": [
    "plt.rcParams['pdf.fonttype'] = 42\n",
    "plt.rcParams['svg.fonttype'] = 'none'\n",
    "\n",
    "sc.set_figure_params(figsize=(6, 6), frameon=False)\n",
    "sns.set_theme(style=\"white\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "19b8f4a8-687a-40dc-b03b-43be3e50024d",
   "metadata": {},
   "source": [
    "## Recap: the <span style=\"font-variant:small-caps; font-size: 28px\">Heimdall</span> framework"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "544a3d79-e3c8-44b6-948c-c90fc1eb00c4",
   "metadata": {},
   "source": [
    "We are interested in implementing various popular tokenizers for single-cell foundation models within a standardized framework, which enables us to isolate the impact of the tokenizer _itself_ on downstream performance."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "11ed0c17-4bcb-450d-9b9d-ff4975782441",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-11-18T08:29:41.535566Z",
     "iopub.status.busy": "2025-11-18T08:29:41.535374Z",
     "iopub.status.idle": "2025-11-18T08:29:41.647645Z",
     "shell.execute_reply": "2025-11-18T08:29:41.647097Z",
     "shell.execute_reply.started": "2025-11-18T08:29:41.535551Z"
    }
   },
   "source": [
    "<figure>\n",
    "  <img src=\"../_static/heimdall_framework.png\"/>\n",
    "  <figcaption><b>Fig 1A.</b> Modular conceptualization of scFMs in <span style=\"font-variant:small-caps; font-size: 16px\">Heimdall</span>. Gene identities and expression values from each single-cell input\n",
    "are processed by a cell tokenization scheme (tokenizer) that generates a sequence-based “cell sentence”. The tokenizer is decomposed into three modules: a gene identity encoder ($F_\\textbf{G}$), an expression encoder ($F_\\textbf{E}$), and a cell constructor ($F_\\textbf{C}$). The tokenizer output is then passed to a sequence-based model (e.g., a transformer).</figcaption>\n",
    "</figure>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9f9e3fca-6004-415a-b06a-c72fceb6e3fb",
   "metadata": {},
   "source": [
    "## Experiment setup (via `hydra`)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "97c45960-25f5-4560-9f42-6c043af36e75",
   "metadata": {},
   "source": [
    "We use [`hydra`](https://github.com/facebookresearch/hydra) to configure each <span style=\"font-variant:small-caps; font-size: 16px\">Heimdall</span> run. First, we detail the configuration files that are shared across all tokenizers for this cross-tissue generalization experiment."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "982a7176-dc26-4052-b335-836629b2c030",
   "metadata": {},
   "source": [
    "Let's take a closer look at some of the most important configs..."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0ba03e03-37e3-4153-8fbe-72ed3169869e",
   "metadata": {},
   "source": [
    "### `experiments`\n",
    "Top-level experiment config that specifies other essential configs."
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "fa5ebc1d-772b-4f62-8561-f013a1fe1ade",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "\n",
    "```yaml\n",
    "defaults:\n",
    "  - override /dataset: new_sctab  # dataset for the relevant task\n",
    "  - override /tasks: new_sctab_split  # dataset for the relevant task\n",
    "  - override /model: transformer  # the chosen model\n",
    "  - override /scheduler: cosine  # contains scheduler\n",
    "  - override /trainer: default  # contains optimizer and trainer details\n",
    "  - override /optimizer: AdamW\n",
    "  - override /fg: random\n",
    "  - override /fe: noop\n",
    "  - override /fc: geneformer\n",
    "  \n",
    "seed: 55 # random seed for reproducibility\n",
    "\n",
    "project_name: new_sctab_split1  # project name for WandB\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "55fa42d4-b286-4442-86c1-41b401fddd2d",
   "metadata": {},
   "source": [
    "### `dataset`\n",
    "Specifies the path to the dataset, as well as preprocessing arguments."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1701403c-84c7-4229-a028-ac0811307131",
   "metadata": {},
   "source": [
    "```yaml\n",
    "dataset_name: new_sctab\n",
    "\n",
    "preprocess_args:\n",
    "  data_path: ${data_path}/sctab/tissue_splits_spencer/scTab_GItract_train.h5ad\n",
    "  top_n_genes: false\n",
    "  normalize: true\n",
    "  log_1p: true\n",
    "  scale_data: false\n",
    "  species: human\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b8344261-f30c-4786-b00b-65508d3a6d83",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-11-18T09:39:12.169251Z",
     "iopub.status.busy": "2025-11-18T09:39:12.169040Z",
     "iopub.status.idle": "2025-11-18T09:39:12.172309Z",
     "shell.execute_reply": "2025-11-18T09:39:12.171691Z",
     "shell.execute_reply.started": "2025-11-18T09:39:12.169232Z"
    }
   },
   "source": [
    "### `task`\n",
    "Specifies the task to use for training the model, including training metrics, dataset splits, the architecture of the task \"head\" (used for predicting the task outputs), and the loss function. In this case, the predefined dataset splits (stored in the `adata.obs[\"split3\"]` column) reserve gastrointestinal tract cells exclusively for training, while brain cells are used exclusively from validation/testing."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "37ee2aab-b541-4023-8f79-35519c5eac0d",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-11-18T09:39:34.703102Z",
     "iopub.status.busy": "2025-11-18T09:39:34.702906Z",
     "iopub.status.idle": "2025-11-18T09:39:34.706430Z",
     "shell.execute_reply": "2025-11-18T09:39:34.705788Z",
     "shell.execute_reply.started": "2025-11-18T09:39:34.703087Z"
    }
   },
   "source": [
    "```yaml\n",
    "type: Heimdall.task.SingleInstanceTask\n",
    "\n",
    "args:\n",
    "  task_type: multiclass\n",
    "  label_col_name: cell_type\n",
    "  metrics: [Accuracy, MatthewsCorrCoef, ConfusionMatrix]\n",
    "  track_metric: MatthewsCorrCoef\n",
    "  splits:\n",
    "    type: predefined\n",
    "    col: split3\n",
    "    keys_:\n",
    "      train: train\n",
    "      val: val\n",
    "      test: test\n",
    "  early_stopping_patience: 5\n",
    "  early_stopping: true\n",
    "  shuffle: true\n",
    "  batchsize: 32\n",
    "  epochs: 50\n",
    "  dataset_config:\n",
    "    type: Heimdall.datasets.SingleInstanceDataset\n",
    "  head_config:\n",
    "    type: Heimdall.models.LinearCellPredHead\n",
    "    args:\n",
    "  loss_config:\n",
    "    type: Heimdall.losses.FlattenCrossEntropyLoss\n",
    "\n",
    "cell_rep_config:\n",
    "  type: Heimdall.cell_representations.CellRepresentation\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5f656626-7fc7-444a-a0cc-70936417179a",
   "metadata": {},
   "source": [
    "### `model`\n",
    "Specifies the model architecture used for the scFM."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "acc652da-b293-40da-b010-8316d85956c6",
   "metadata": {},
   "source": [
    "```yaml\n",
    "type: Heimdall.models.Transformer\n",
    "name: transformer\n",
    "\n",
    "args:\n",
    "  d_model: 128\n",
    "  pos_enc: BERT\n",
    "  num_encoder_layers: 2\n",
    "  nhead: 4\n",
    "  hidden_act: gelu\n",
    "  hidden_dropout_prob: 0.1\n",
    "  use_flash_attn: false\n",
    "  pooling: cls_pooling # or \"mean_pooling\"\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6adbc034-f25e-4238-b4ae-eaa51bfa8863",
   "metadata": {},
   "source": [
    "## Modular reimplementation of the Geneformer tokenizer"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c7b532ba-6b90-4566-bc03-91dafd45f342",
   "metadata": {},
   "source": [
    "Having configured everything except the tokenizer, we now focus on implementing the tokenizer. For practice, let's implement the [Geneformer](https://www.nature.com/articles/s41586-023-06139-9) tokenizer."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5171b2b8-b0af-471e-9a1c-fccf6bb5c9f7",
   "metadata": {},
   "source": [
    "### `fg` - the gene identity encoder ($F_\\textbf{G}$)\n",
    "Specifies the `Fg` implementation for this tokenizer, as well as the `torch.nn.Module` used for providing the embeddings of the genes. Here, we use the `random` implementation, which assigns a randomly-initialized embedding vector of dimensionality `model.args.d_model` to each gene in the cell."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c7b9209a-7fa1-4667-ad17-fda81ac81ad5",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-11-18T10:54:20.100098Z",
     "iopub.status.busy": "2025-11-18T10:54:20.099902Z",
     "iopub.status.idle": "2025-11-18T10:54:20.103186Z",
     "shell.execute_reply": "2025-11-18T10:54:20.102651Z",
     "shell.execute_reply.started": "2025-11-18T10:54:20.100083Z"
    }
   },
   "source": [
    "```yaml\n",
    "type: Heimdall.fg.IdentityFg\n",
    "\n",
    "args:\n",
    "  embedding_parameters:\n",
    "    type: Heimdall.embedding.FlexibleTypeEmbedding\n",
    "    args:\n",
    "      num_embeddings: vocab_size\n",
    "      embedding_dim: ${fg.args.d_embedding}\n",
    "  d_embedding: ${model.args.d_model}\n",
    "  frozen: false\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "812a0312-aa19-45eb-a317-2c3d1b903a83",
   "metadata": {},
   "source": [
    "### `fe` - the gene expression encoder ($F_\\textbf{E}$)\n",
    "Specifies the `Fe` implementation for this tokenizer, as well as the `torch.nn.Module` used for providing the embeddings of the genes' expression levels. Here, we use the `noop` implementation, which simply outputs a vector of zeros of dimensionality `model.args.d_model` for each gene in the cell, regardless of the gene's expression level."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "15d4976e-8078-4720-ab24-1bd1089e05bd",
   "metadata": {},
   "source": [
    "```yaml\n",
    "type: Heimdall.fe.IdentityFe\n",
    "name: Heimdall.fe.IdentityFe\n",
    "\n",
    "args:\n",
    "  embedding_parameters:\n",
    "    type: Heimdall.embedding.ZeroBroadcast\n",
    "    args:\n",
    "      out_features: ${fe.args.d_embedding}\n",
    "  d_embedding: ${model.args.d_model}\n",
    "  drop_zeros: true\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2d0fee0e-5f0d-49dc-82db-dca172276121",
   "metadata": {},
   "source": [
    "### `fc` - the single-cell representation function ($F_\\textbf{C}$)\n",
    "Specifies the "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "76464f13-de45-428a-aeb5-9926dc7c36df",
   "metadata": {},
   "source": [
    "```yaml\n",
    "type: Heimdall.fc.Fc\n",
    "\n",
    "args:\n",
    "  max_input_length: 2048\n",
    "  embedding_parameters:\n",
    "    type: torch.nn.Module  # Should throw an error if called\n",
    "  tailor_config:\n",
    "    type: Heimdall.tailor.ReorderTailor\n",
    "  order_config:\n",
    "    type: Heimdall.order.ExpressionOrder\n",
    "  reduce_config:\n",
    "    type: Heimdall.reduce.SumReduce\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ed499448-7af9-430c-97b8-9dfa04faa2ac",
   "metadata": {},
   "source": [
    "## Putting it all together"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "c2e1ac94-9934-41be-af19-d514e78d46be",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-11-18T11:48:41.499147Z",
     "iopub.status.busy": "2025-11-18T11:48:41.498946Z",
     "iopub.status.idle": "2025-11-18T11:48:41.614303Z",
     "shell.execute_reply": "2025-11-18T11:48:41.613603Z",
     "shell.execute_reply.started": "2025-11-18T11:48:41.499131Z"
    }
   },
   "outputs": [],
   "source": [
    "from omegaconf import OmegaConf\n",
    "with hydra.initialize(version_base=None, config_path=\"../Heimdall/config\"):\n",
    "    config = hydra.compose(\n",
    "        config_name=\"config\",\n",
    "        overrides=[\n",
    "            \"+experiments=sctab_split1_all\",\n",
    "            \"fg=random\",\n",
    "            \"fe=noop\",\n",
    "            \"fc=geneformer\",\n",
    "        ],\n",
    "    )\n",
    "\n",
    "    OmegaConf.resolve(config)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "800e06fc-c7d7-4f2a-a714-c688a1e14a3c",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-11-18T11:51:04.617168Z",
     "iopub.status.busy": "2025-11-18T11:51:04.616904Z",
     "iopub.status.idle": "2025-11-18T11:51:04.628152Z",
     "shell.execute_reply": "2025-11-18T11:51:04.627324Z",
     "shell.execute_reply.started": "2025-11-18T11:51:04.617145Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "project_name: new_sctab_split1\n",
      "run_name: Heimdall.fg.IdentityFg_Heimdall.fe.IdentityFe_Heimdall.fc.Fc_Heimdall.models.Transformer_lr0.002_bz32\n",
      "work_dir: new_sctab_split1_results/Heimdall.fg.IdentityFg_Heimdall.fe.IdentityFe_Heimdall.fc.Fc_new_sctab_lr0.002_bz32_seed55_agTrue\n",
      "run_wandb: true\n",
      "float_dtype: float32\n",
      "seed: 55\n",
      "data_path: /work/magroup/shared/Heimdall/data\n",
      "ensembl_dir: /work/magroup/shared/Heimdall/data\n",
      "cache_preprocessed_dataset_dir: /scratch/heimdall/shared/cache\n",
      "entity: Heimdall\n",
      "only_preprocess_data: false\n",
      "model:\n",
      "  type: Heimdall.models.Transformer\n",
      "  name: transformer\n",
      "  args:\n",
      "    d_model: 128\n",
      "    pos_enc: BERT\n",
      "    num_encoder_layers: 2\n",
      "    nhead: 4\n",
      "    hidden_act: gelu\n",
      "    hidden_dropout_prob: 0.1\n",
      "    use_flash_attn: false\n",
      "    pooling: cls_pooling\n",
      "dataset:\n",
      "  dataset_name: new_sctab\n",
      "  preprocess_args:\n",
      "    data_path: /work/magroup/shared/Heimdall/data/sctab/tissue_splits_spencer/scTab_GItract_train.h5ad\n",
      "    top_n_genes: false\n",
      "    normalize: true\n",
      "    log_1p: true\n",
      "    scale_data: false\n",
      "    species: human\n",
      "tasks:\n",
      "  type: Heimdall.task.SingleInstanceTask\n",
      "  args:\n",
      "    task_type: multiclass\n",
      "    label_col_name: cell_type\n",
      "    metrics:\n",
      "    - Accuracy\n",
      "    - MatthewsCorrCoef\n",
      "    - ConfusionMatrix\n",
      "    track_metric: MatthewsCorrCoef\n",
      "    splits:\n",
      "      type: predefined\n",
      "      col: split1\n",
      "      keys_:\n",
      "        train: train\n",
      "        val: val\n",
      "        test: test\n",
      "    early_stopping_patience: 5\n",
      "    early_stopping: true\n",
      "    shuffle: true\n",
      "    batchsize: 32\n",
      "    epochs: 50\n",
      "    dataset_config:\n",
      "      type: Heimdall.datasets.SingleInstanceDataset\n",
      "    head_config:\n",
      "      type: Heimdall.models.LinearCellPredHead\n",
      "      args: null\n",
      "    loss_config:\n",
      "      type: Heimdall.losses.FlattenCrossEntropyLoss\n",
      "  cell_rep_config:\n",
      "    type: Heimdall.cell_representations.CellRepresentation\n",
      "scheduler:\n",
      "  name: cosine\n",
      "  lr_schedule_type: cosine\n",
      "  warmup_ratio: 0.1\n",
      "  num_epochs: 20\n",
      "trainer:\n",
      "  type: Heimdall.trainer.HeimdallTrainer\n",
      "  args:\n",
      "    random_seed: 55\n",
      "    accumulate_grad_batches: 1\n",
      "    grad_norm_clip: 1.0\n",
      "    fastdev: false\n",
      "    skip_umaps: false\n",
      "  cpu: false\n",
      "optimizer:\n",
      "  name: AdamW\n",
      "  args:\n",
      "    lr: 0.002\n",
      "    weight_decay: 0.1\n",
      "    betas:\n",
      "    - 0.9\n",
      "    - 0.95\n",
      "    foreach: false\n",
      "fc:\n",
      "  type: Heimdall.fc.Fc\n",
      "  args:\n",
      "    max_input_length: 2048\n",
      "    embedding_parameters:\n",
      "      type: torch.nn.Module\n",
      "    tailor_config:\n",
      "      type: Heimdall.tailor.ReorderTailor\n",
      "    order_config:\n",
      "      type: Heimdall.order.ExpressionOrder\n",
      "    reduce_config:\n",
      "      type: Heimdall.reduce.SumReduce\n",
      "fe:\n",
      "  type: Heimdall.fe.IdentityFe\n",
      "  name: Heimdall.fe.IdentityFe\n",
      "  args:\n",
      "    embedding_parameters:\n",
      "      type: Heimdall.embedding.ZeroBroadcast\n",
      "      args:\n",
      "        out_features: 128\n",
      "    d_embedding: 128\n",
      "    drop_zeros: true\n",
      "fg:\n",
      "  type: Heimdall.fg.IdentityFg\n",
      "  args:\n",
      "    embedding_parameters:\n",
      "      type: Heimdall.embedding.FlexibleTypeEmbedding\n",
      "      args:\n",
      "        num_embeddings: vocab_size\n",
      "        embedding_dim: 128\n",
      "    d_embedding: 128\n",
      "    frozen: false\n",
      "loss:\n",
      "  type: Heimdall.losses.FlattenCrossEntropyLoss\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print(OmegaConf.to_yaml(config))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dd499e6d-8dde-4091-9f73-6c42e5c617df",
   "metadata": {},
   "source": [
    "## Training the model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "73e975cb-04c3-4bfa-bca4-088903f526dd",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-11-25T14:40:28.001170Z",
     "iopub.status.busy": "2025-11-25T14:40:28.000984Z",
     "iopub.status.idle": "2025-11-25T14:43:23.312257Z",
     "shell.execute_reply": "2025-11-25T14:43:23.311173Z",
     "shell.execute_reply.started": "2025-11-25T14:40:28.001152Z"
    }
   },
   "outputs": [],
   "source": [
    "from Heimdall.trainer import setup_trainer\n",
    "def training_loop(config): \n",
    "    trainer = setup_trainer(config, cpu=config.trainer.cpu)\n",
    "    if trainer is not None:\n",
    "        trainer.fit()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "86e4a463-3930-4dfd-b303-135891b9419c",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-11-18T11:51:25.244542Z",
     "iopub.status.busy": "2025-11-18T11:51:25.244340Z",
     "iopub.status.idle": "2025-11-18T11:51:25.246846Z",
     "shell.execute_reply": "2025-11-18T11:51:25.246281Z",
     "shell.execute_reply.started": "2025-11-18T11:51:25.244526Z"
    }
   },
   "outputs": [],
   "source": [
    "from accelerate import notebook_launcher"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1da8625f-40f5-4be9-899c-77663ccb0711",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "args = (config,)\n",
    "notebook_launcher(training_loop, args, num_processes=1)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:flash_attn_py312]",
   "language": "python",
   "name": "conda-env-flash_attn_py312-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
