{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "b69fdc71-5ff0-49a1-83b0-037a4c266468",
   "metadata": {},
   "source": [
    "# Cell type annotation dataset processing notebook\n",
    "\n",
    "This notebook contains processing scripts for each dataset used for cell type annotation.\n",
    "\n",
    "First, run the header block below. Then, you can run any one of the data processing block without runing the full notebook."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "2ed4162c-2d79-44e0-8f89-fc8eed3fbb52",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-09-09T04:29:49.008176Z",
     "iopub.status.busy": "2024-09-09T04:29:49.007852Z",
     "iopub.status.idle": "2024-09-09T04:29:49.012405Z",
     "shell.execute_reply": "2024-09-09T04:29:49.011461Z",
     "shell.execute_reply.started": "2024-09-09T04:29:49.008142Z"
    }
   },
   "outputs": [],
   "source": [
    "### Uncomment and run this block to install necessary packages\n",
    "# pip install -U anndata scanpy gdown pandas tqdm mygene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "816c363e-c45f-4ff3-9f41-60db5ecd16dc",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-09-09T04:30:02.889048Z",
     "iopub.status.busy": "2024-09-09T04:30:02.888389Z",
     "iopub.status.idle": "2024-09-09T04:30:02.895295Z",
     "shell.execute_reply": "2024-09-09T04:30:02.893650Z",
     "shell.execute_reply.started": "2024-09-09T04:30:02.888989Z"
    }
   },
   "outputs": [],
   "source": [
    "### Dev only\n",
    "# %load_ext autoreload\n",
    "# %aimport data_processing_utils\n",
    "# %autoreload 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "ec99f2e5-fa6f-4bee-ae1a-d34e8872bc01",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-09-09T04:31:22.939313Z",
     "iopub.status.busy": "2024-09-09T04:31:22.938604Z",
     "iopub.status.idle": "2024-09-09T04:31:47.834786Z",
     "shell.execute_reply": "2024-09-09T04:31:47.833484Z",
     "shell.execute_reply.started": "2024-09-09T04:31:22.939254Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Raw data will be saved to: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw\n",
      "Processed data will be saved to: /work/magroup/shared/Heimdall/data/cell_type_annotation/processed\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/work/magroup/shahula/bin/miniconda3/envs/heimdall/lib/python3.10/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html\n",
      "  from .autonotebook import tqdm as notebook_tqdm\n"
     ]
    }
   ],
   "source": [
    "# Header block\n",
    "import os\n",
    "from pathlib import Path\n",
    "\n",
    "import anndata as ad\n",
    "import gdown\n",
    "import mygene\n",
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "from sklearn.preprocessing import LabelEncoder\n",
    "\n",
    "from Heimdall.data_processing_utils import symbol_to_ensembl_from_ensembl\n",
    "\n",
    "### Change project dir here ###\n",
    "# PROJECT_DIR = Path().resolve().parent  # use the project repository root dir as the project dir\n",
    "PROJECT_DIR = Path(\"/work/magroup/shared/Heimdall\")\n",
    "#################################\n",
    "\n",
    "DATA_DIR = PROJECT_DIR / \"data\" / \"cell_type_annotation\"\n",
    "RAW_DATA_DIR = DATA_DIR / \"raw\"\n",
    "PROCESSED_DATA_DIR = DATA_DIR / \"processed\"\n",
    "\n",
    "os.makedirs(RAW_DATA_DIR, exist_ok=True)\n",
    "os.makedirs(PROCESSED_DATA_DIR, exist_ok=True)\n",
    "\n",
    "print(f\"Raw data will be saved to: {RAW_DATA_DIR}\")\n",
    "print(f\"Processed data will be saved to: {PROCESSED_DATA_DIR}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8e9686fd-abaf-47f5-b165-b860a7a2f22c",
   "metadata": {},
   "source": [
    "## Pancreas\n",
    "\n",
    "https://openproblems.bio/results/label_projection/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "f9921b2f-0442-4a79-88f2-45f22cf939e7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--2024-07-10 11:05:08--  https://ndownloader.figshare.com/files/36086813\n",
      "Resolving ndownloader.figshare.com (ndownloader.figshare.com)... 54.78.158.183, 54.194.168.157, 34.250.28.103, ...\n",
      "Connecting to ndownloader.figshare.com (ndownloader.figshare.com)|54.78.158.183|:443... connected.\n",
      "HTTP request sent, awaiting response... 302 Found\n",
      "Location: https://s3-eu-west-1.amazonaws.com/pfigshare-u-files/36086813/pancreas.h5ad?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIYCQYOYV5JSSROOA/20240710/eu-west-1/s3/aws4_request&X-Amz-Date=20240710T150509Z&X-Amz-Expires=10&X-Amz-SignedHeaders=host&X-Amz-Signature=fad9479828cec518f80fb31951c85a22f5ce3d5e240699af12021b9c067f2ae5 [following]\n",
      "--2024-07-10 11:05:09--  https://s3-eu-west-1.amazonaws.com/pfigshare-u-files/36086813/pancreas.h5ad?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIYCQYOYV5JSSROOA/20240710/eu-west-1/s3/aws4_request&X-Amz-Date=20240710T150509Z&X-Amz-Expires=10&X-Amz-SignedHeaders=host&X-Amz-Signature=fad9479828cec518f80fb31951c85a22f5ce3d5e240699af12021b9c067f2ae5\n",
      "Resolving s3-eu-west-1.amazonaws.com (s3-eu-west-1.amazonaws.com)... 52.218.120.184, 52.218.1.99, 52.218.98.43, ...\n",
      "Connecting to s3-eu-west-1.amazonaws.com (s3-eu-west-1.amazonaws.com)|52.218.120.184|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 892650076 (851M) [application/octet-stream]\n",
      "Saving to: ‘/work/magroup/shared/Heimdall/data/cell_type_annotation/raw/pancreas.h5ad’\n",
      "\n",
      "100%[======================================>] 892,650,076 30.4MB/s   in 30s    \n",
      "\n",
      "2024-07-10 11:05:39 (28.6 MB/s) - ‘/work/magroup/shared/Heimdall/data/cell_type_annotation/raw/pancreas.h5ad’ saved [892650076/892650076]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "!wget https://ndownloader.figshare.com/files/36086813 -O {RAW_DATA_DIR}/pancreas.h5ad"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "6a2cc82c-0fb0-48b9-b589-acc660ea94cf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 16382 × 19093\n",
       "    obs: 'tech', 'celltype', 'size_factors'\n",
       "    layers: 'counts'"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load raw data\n",
    "data = ad.read_h5ad(RAW_DATA_DIR / \"pancreas.h5ad\")\n",
    "data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "c54a92e4-d0af-4d7f-9115-dbd00ec04c02",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 16382 × 15309\n",
       "    obs: 'tech', 'celltype', 'size_factors'\n",
       "    var: 'n_cells'\n",
       "    layers: 'counts'"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Filtering\n",
    "sc.pp.filter_genes(data, min_cells=100, inplace=True)\n",
    "data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "66fe6143-27ea-4bad-8bf7-70ec4f3abe04",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Mapping data directory: /work/magroup/shared/Heimdall/data/gene_mapping/ensembl/human\n",
      "Loading mapping from cache: /work/magroup/shared/Heimdall/data/gene_mapping/ensembl/human/symbol_to_ensembl.json\n",
      "Successfully mapped 14,437 out of 15,309 genes (94.3%)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>n_cells</th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>gene_ensembl</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000121410</th>\n",
       "      <td>3178</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>ENSG00000121410</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000148584</th>\n",
       "      <td>5937</td>\n",
       "      <td>A1CF</td>\n",
       "      <td>ENSG00000148584</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000175899</th>\n",
       "      <td>789</td>\n",
       "      <td>A2M</td>\n",
       "      <td>ENSG00000175899</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000166535</th>\n",
       "      <td>487</td>\n",
       "      <td>A2ML1</td>\n",
       "      <td>ENSG00000166535</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000128274</th>\n",
       "      <td>879</td>\n",
       "      <td>A4GALT</td>\n",
       "      <td>ENSG00000128274</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000070476</th>\n",
       "      <td>5128</td>\n",
       "      <td>ZXDC</td>\n",
       "      <td>ENSG00000070476</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000162378</th>\n",
       "      <td>6073</td>\n",
       "      <td>ZYG11B</td>\n",
       "      <td>ENSG00000162378</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000159840</th>\n",
       "      <td>5334</td>\n",
       "      <td>ZYX</td>\n",
       "      <td>ENSG00000159840</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000074755</th>\n",
       "      <td>5093</td>\n",
       "      <td>ZZEF1</td>\n",
       "      <td>ENSG00000074755</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000036549</th>\n",
       "      <td>4694</td>\n",
       "      <td>ZZZ3</td>\n",
       "      <td>ENSG00000036549</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>15309 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 n_cells gene_symbol     gene_ensembl\n",
       "ENSG00000121410     3178        A1BG  ENSG00000121410\n",
       "ENSG00000148584     5937        A1CF  ENSG00000148584\n",
       "ENSG00000175899      789         A2M  ENSG00000175899\n",
       "ENSG00000166535      487       A2ML1  ENSG00000166535\n",
       "ENSG00000128274      879      A4GALT  ENSG00000128274\n",
       "...                  ...         ...              ...\n",
       "ENSG00000070476     5128        ZXDC  ENSG00000070476\n",
       "ENSG00000162378     6073      ZYG11B  ENSG00000162378\n",
       "ENSG00000159840     5334         ZYX  ENSG00000159840\n",
       "ENSG00000074755     5093       ZZEF1  ENSG00000074755\n",
       "ENSG00000036549     4694        ZZZ3  ENSG00000036549\n",
       "\n",
       "[15309 rows x 3 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Gene id mapping\n",
    "symbol_to_ensembl_mapping = symbol_to_ensembl_from_ensembl(\n",
    "    data_dir=PROJECT_DIR / \"data\", genes=data.var.index.tolist(), species=\"human\")\n",
    "data.uns[\"gene_mapping:symbol_to_ensembl\"] = symbol_to_ensembl_mapping.mapping_full\n",
    "\n",
    "data.var[\"gene_symbol\"] = data.var.index\n",
    "data.var[\"gene_ensembl\"] = data.var[\"gene_symbol\"].map(symbol_to_ensembl_mapping.mapping_combined.get)\n",
    "data.var.index = data.var.index.map(symbol_to_ensembl_mapping.mapping_reduced)\n",
    "\n",
    "data.var"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "1cecc70d-6b07-45eb-8d45-df65b571d9a3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 16382 × 15309\n",
       "    obs: 'tech', 'celltype', 'size_factors', 'species', 'task_celltype', 'batch'\n",
       "    var: 'n_cells', 'gene_symbol', 'gene_ensembl'\n",
       "    uns: 'gene_mapping:symbol_to_ensembl', 'celltype_order', 'batch_order'\n",
       "    layers: 'counts'"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Standardize attributes\n",
    "data.obs[\"species\"] = \"Homo sapiens\"\n",
    "le = LabelEncoder()\n",
    "\n",
    "data.obs[\"task_celltype\"] = le.fit_transform(data.obs[\"celltype\"])\n",
    "data.uns[\"celltype_order\"] = le.classes_.tolist()\n",
    "\n",
    "data.obs[\"batch\"] = le.fit_transform(data.obs[\"tech\"])\n",
    "data.uns[\"batch_order\"] = le.classes_.tolist()\n",
    "\n",
    "data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "2a271ec3-d417-4d5b-bf24-2130652ed733",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save processed data\n",
    "data.write_h5ad(PROCESSED_DATA_DIR / \"pancreas.h5ad\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5708e020-7838-4acd-bba3-52910899bea6",
   "metadata": {},
   "source": [
    "## Zheng68k"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "ec8cbdfb-0c0c-4f29-9e35-2f9cd351f760",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--2024-07-10 11:07:12--  https://cf.10xgenomics.com/samples/cell-exp/1.1.0/fresh_68k_pbmc_donor_a/fresh_68k_pbmc_donor_a_filtered_gene_bc_matrices.tar.gz\n",
      "Resolving cf.10xgenomics.com (cf.10xgenomics.com)... 104.18.1.173, 104.18.0.173, 2606:4700::6812:ad, ...\n",
      "Connecting to cf.10xgenomics.com (cf.10xgenomics.com)|104.18.1.173|:443... connected.\n",
      "WARNING: cannot verify cf.10xgenomics.com's certificate, issued by ‘/C=US/O=Let's Encrypt/CN=E1’:\n",
      "  Issued certificate has expired.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 124442812 (119M) [application/x-tar]\n",
      "Saving to: ‘/work/magroup/shared/Heimdall/data/cell_type_annotation/raw/zheng68k.tar.gz’\n",
      "\n",
      "100%[======================================>] 124,442,812 77.7MB/s   in 1.5s   \n",
      "\n",
      "2024-07-10 11:07:14 (77.7 MB/s) - ‘/work/magroup/shared/Heimdall/data/cell_type_annotation/raw/zheng68k.tar.gz’ saved [124442812/124442812]\n",
      "\n",
      "filtered_matrices_mex/\n",
      "filtered_matrices_mex/hg19/\n",
      "filtered_matrices_mex/hg19/barcodes.tsv\n",
      "filtered_matrices_mex/hg19/genes.tsv\n",
      "filtered_matrices_mex/hg19/matrix.mtx\n",
      "--2024-07-10 11:07:25--  https://github.com/10XGenomics/single-cell-3prime-paper/raw/989aeed58745e01fe13acc439bdc19c2c185a1aa/pbmc68k_analysis/68k_pbmc_barcodes_annotation.tsv\n",
      "Resolving github.com (github.com)... 140.82.112.3\n",
      "Connecting to github.com (github.com)|140.82.112.3|:443... connected.\n",
      "HTTP request sent, awaiting response... 302 Found\n",
      "Location: https://raw.githubusercontent.com/10XGenomics/single-cell-3prime-paper/989aeed58745e01fe13acc439bdc19c2c185a1aa/pbmc68k_analysis/68k_pbmc_barcodes_annotation.tsv [following]\n",
      "--2024-07-10 11:07:26--  https://raw.githubusercontent.com/10XGenomics/single-cell-3prime-paper/989aeed58745e01fe13acc439bdc19c2c185a1aa/pbmc68k_analysis/68k_pbmc_barcodes_annotation.tsv\n",
      "Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.108.133, 185.199.109.133, 185.199.111.133, ...\n",
      "Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.108.133|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 4800488 (4.6M) [text/plain]\n",
      "Saving to: ‘/work/magroup/shared/Heimdall/data/cell_type_annotation/raw/zheng68k/annotation.tsv’\n",
      "\n",
      "100%[======================================>] 4,800,488   --.-K/s   in 0.06s   \n",
      "\n",
      "2024-07-10 11:07:26 (77.1 MB/s) - ‘/work/magroup/shared/Heimdall/data/cell_type_annotation/raw/zheng68k/annotation.tsv’ saved [4800488/4800488]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "!wget https://cf.10xgenomics.com/samples/cell-exp/1.1.0/fresh_68k_pbmc_donor_a/fresh_68k_pbmc_donor_a_filtered_gene_bc_matrices.tar.gz --no-check-certificate -O {RAW_DATA_DIR}/zheng68k.tar.gz\n",
    "!tar -xzvf {RAW_DATA_DIR}/zheng68k.tar.gz -C {RAW_DATA_DIR} && mv {RAW_DATA_DIR}/filtered_matrices_mex/hg19 {RAW_DATA_DIR}/zheng68k\n",
    "\n",
    "!wget https://github.com/10XGenomics/single-cell-3prime-paper/raw/989aeed58745e01fe13acc439bdc19c2c185a1aa/pbmc68k_analysis/68k_pbmc_barcodes_annotation.tsv -O {RAW_DATA_DIR}/zheng68k/annotation.tsv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "c2387c03-d636-4781-aeef-59952ed19e6a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 68579 × 32738\n",
       "    obs: 'TSNE.1', 'TSNE.2', 'celltype'\n",
       "    var: 'Symbol'"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Consolidate anndata\n",
    "data = ad.read_mtx(RAW_DATA_DIR / \"zheng68k\" / \"matrix.mtx\").T\n",
    "\n",
    "barcodes = pd.read_csv(RAW_DATA_DIR / \"zheng68k\" / \"barcodes.tsv\", header=None)\n",
    "genes = pd.read_csv(RAW_DATA_DIR / \"zheng68k\" / \"genes.tsv\", sep=\"\\t\", header=None, names=[\"Symbol\"], index_col=0)\n",
    "\n",
    "annot = pd.read_csv(RAW_DATA_DIR / \"zheng68k\" / \"annotation.tsv\", sep=\"\\t\").set_index(\"barcodes\").reindex(barcodes[0])\n",
    "annot.index.name = None\n",
    "\n",
    "data.obs = annot\n",
    "data.var = genes\n",
    "\n",
    "data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "deb4e56e-2f85-4d43-a356-5d9701d15858",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 68579 × 11240\n",
       "    obs: 'TSNE.1', 'TSNE.2', 'celltype'\n",
       "    var: 'Symbol', 'n_cells'"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Filtering\n",
    "sc.pp.filter_genes(data, min_cells=100, inplace=True)\n",
    "data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "e5b49047-f85c-477c-af20-0615e3092a2a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 68579 × 11240\n",
       "    obs: 'TSNE.1', 'TSNE.2', 'celltype', 'species', 'task_celltype'\n",
       "    var: 'gene_symbol', 'n_cells', 'gene_ensembl'\n",
       "    uns: 'celltype_order'"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Standardize attributes\n",
    "data.obs[\"species\"] = \"Homo sapiens\"\n",
    "le = LabelEncoder()\n",
    "\n",
    "data.obs[\"task_celltype\"] = le.fit_transform(data.obs[\"celltype\"])\n",
    "data.uns[\"celltype_order\"] = le.classes_.tolist()\n",
    "\n",
    "data.var.rename(columns={\"Symbol\": \"gene_symbol\"}, inplace=True)\n",
    "data.var[\"gene_ensembl\"] = data.var.index\n",
    "\n",
    "data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "8ec2f56c-7359-4514-8a7f-afc8e1f11a88",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save processed data\n",
    "data.write_h5ad(PROCESSED_DATA_DIR / \"zheng68k.h5ad\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "420c1113-d35c-45c8-9fc6-ac16a79c4990",
   "metadata": {},
   "source": [
    "## Multiple Sclerosis (MS)\n",
    "\n",
    "https://github.com/bowang-lab/scGPT/tree/main/data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "cd1b4fa1-1a13-4a92-9de7-db6f53a752b3",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Downloading...\n",
      "From: https://drive.google.com/uc?id=1casFhq4InuBNhJLMnGebzkRXM2UTTeQG\n",
      "To: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/ms.h5ad\n",
      "100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 47.1M/47.1M [00:00<00:00, 89.2MB/s]\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'/work/magroup/shared/Heimdall/data/cell_type_annotation/raw/ms.h5ad'"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gdown.download(\n",
    "    url=\"https://drive.google.com/file/d/1casFhq4InuBNhJLMnGebzkRXM2UTTeQG/view?usp=drive_link\",\n",
    "    output=str(RAW_DATA_DIR / \"ms.h5ad\"),\n",
    "    fuzzy=True,\n",
    ")\n",
    "\n",
    "# gdown.download(\n",
    "#     url=\"https://drive.google.com/file/d/1bV1SHKVZgkcL-RmmuN51_IIUJTSJbXOi/view?usp=drive_link\",\n",
    "#     output=str(RAW_DATA_DIR / \"ms_c.h5ad\"),\n",
    "#     fuzzy=True,\n",
    "# )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "77fb6f31-7fda-4c38-b48d-4cdb47af41cb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 13468 × 3000\n",
       "    obs: 'Sample Characteristic[organism]', 'Sample Characteristic Ontology Term[organism]', 'Sample Characteristic[individual]', 'Sample Characteristic Ontology Term[individual]', 'Sample Characteristic[sex]', 'Sample Characteristic Ontology Term[sex]', 'Sample Characteristic[age]', 'Sample Characteristic Ontology Term[age]', 'Sample Characteristic[developmental stage]', 'Sample Characteristic Ontology Term[developmental stage]', 'Sample Characteristic[organism part]', 'Sample Characteristic Ontology Term[organism part]', 'Sample Characteristic[sampling site]', 'Sample Characteristic Ontology Term[sampling site]', 'Sample Characteristic[disease]', 'Sample Characteristic Ontology Term[disease]', 'Sample Characteristic[organism status]', 'Sample Characteristic Ontology Term[organism status]', 'Sample Characteristic[cause of death]', 'Sample Characteristic Ontology Term[cause of death]', 'Sample Characteristic[clinical history]', 'Sample Characteristic Ontology Term[clinical history]', 'Factor Value[disease]', 'Factor Value Ontology Term[disease]', 'Factor Value[sampling site]', 'Factor Value Ontology Term[sampling site]', 'Factor Value[inferred cell type - ontology labels]', 'Factor Value Ontology Term[inferred cell type - ontology labels]', 'Factor Value[inferred cell type - authors labels]', 'Factor Value Ontology Term[inferred cell type - authors labels]', 'str_batch', 'celltype'\n",
       "    var: 'index_column', 'gene_name'\n",
       "    uns: 'log1p', 'neighbors', 'umap'\n",
       "    obsm: 'X_pca', 'X_umap'\n",
       "    obsp: 'connectivities', 'distances'"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load raw data\n",
    "data = ad.read_h5ad(RAW_DATA_DIR / \"ms.h5ad\")\n",
    "data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "ac6a5804-feae-4935-a279-fff56af0825b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 13468 × 2336\n",
       "    obs: 'Sample Characteristic[organism]', 'Sample Characteristic Ontology Term[organism]', 'Sample Characteristic[individual]', 'Sample Characteristic Ontology Term[individual]', 'Sample Characteristic[sex]', 'Sample Characteristic Ontology Term[sex]', 'Sample Characteristic[age]', 'Sample Characteristic Ontology Term[age]', 'Sample Characteristic[developmental stage]', 'Sample Characteristic Ontology Term[developmental stage]', 'Sample Characteristic[organism part]', 'Sample Characteristic Ontology Term[organism part]', 'Sample Characteristic[sampling site]', 'Sample Characteristic Ontology Term[sampling site]', 'Sample Characteristic[disease]', 'Sample Characteristic Ontology Term[disease]', 'Sample Characteristic[organism status]', 'Sample Characteristic Ontology Term[organism status]', 'Sample Characteristic[cause of death]', 'Sample Characteristic Ontology Term[cause of death]', 'Sample Characteristic[clinical history]', 'Sample Characteristic Ontology Term[clinical history]', 'Factor Value[disease]', 'Factor Value Ontology Term[disease]', 'Factor Value[sampling site]', 'Factor Value Ontology Term[sampling site]', 'Factor Value[inferred cell type - ontology labels]', 'Factor Value Ontology Term[inferred cell type - ontology labels]', 'Factor Value[inferred cell type - authors labels]', 'Factor Value Ontology Term[inferred cell type - authors labels]', 'str_batch', 'celltype'\n",
       "    var: 'index_column', 'gene_name', 'n_cells'\n",
       "    uns: 'log1p', 'neighbors', 'umap'\n",
       "    obsm: 'X_pca', 'X_umap'\n",
       "    obsp: 'connectivities', 'distances'"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Filtering\n",
    "sc.pp.filter_genes(data, min_cells=100, inplace=True)\n",
    "data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "ff7117dd-0912-4a51-a8f7-bbddaa85b343",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 13468 × 2336\n",
       "    obs: 'Sample Characteristic[organism]', 'Sample Characteristic Ontology Term[organism]', 'Sample Characteristic[individual]', 'Sample Characteristic Ontology Term[individual]', 'Sample Characteristic[sex]', 'Sample Characteristic Ontology Term[sex]', 'Sample Characteristic[age]', 'Sample Characteristic Ontology Term[age]', 'Sample Characteristic[developmental stage]', 'Sample Characteristic Ontology Term[developmental stage]', 'Sample Characteristic[organism part]', 'Sample Characteristic Ontology Term[organism part]', 'Sample Characteristic[sampling site]', 'Sample Characteristic Ontology Term[sampling site]', 'Sample Characteristic[disease]', 'Sample Characteristic Ontology Term[disease]', 'Sample Characteristic[organism status]', 'Sample Characteristic Ontology Term[organism status]', 'Sample Characteristic[cause of death]', 'Sample Characteristic Ontology Term[cause of death]', 'Sample Characteristic[clinical history]', 'Sample Characteristic Ontology Term[clinical history]', 'Factor Value[disease]', 'Factor Value Ontology Term[disease]', 'Factor Value[sampling site]', 'Factor Value Ontology Term[sampling site]', 'Factor Value[inferred cell type - ontology labels]', 'Factor Value Ontology Term[inferred cell type - ontology labels]', 'Factor Value[inferred cell type - authors labels]', 'Factor Value Ontology Term[inferred cell type - authors labels]', 'str_batch', 'celltype', 'species', 'task_celltype'\n",
       "    var: 'gene_ensembl', 'gene_symbol', 'n_cells'\n",
       "    uns: 'log1p', 'neighbors', 'umap', 'celltype_order'\n",
       "    obsm: 'X_pca', 'X_umap'\n",
       "    obsp: 'connectivities', 'distances'"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Standardize attributes\n",
    "data.obs[\"species\"] = \"Homo sapiens\"\n",
    "data.var.rename(columns={\"index_column\": \"gene_ensembl\", \"gene_name\": \"gene_symbol\"}, inplace=True)\n",
    "\n",
    "le = LabelEncoder()\n",
    "\n",
    "data.obs[\"task_celltype\"] = le.fit_transform(data.obs[\"celltype\"])\n",
    "data.uns[\"celltype_order\"] = le.classes_.tolist()\n",
    "\n",
    "data.var.rename(columns={\"index_column\": \"gene_ensembl\", \"gene_name\": \"gene_symbo\"}, inplace=True)\n",
    "\n",
    "data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "9f577ad5-feca-49b9-9d7f-1d93a39adecc",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save processed data\n",
    "data.write_h5ad(PROCESSED_DATA_DIR / \"ms.h5ad\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e685789b-955d-4856-af45-8270d8a1ff13",
   "metadata": {},
   "source": [
    "## Myeloid (MYE)\n",
    "\n",
    "https://github.com/bowang-lab/scGPT/tree/main/data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "1ad54df1-0d97-4a7f-b7aa-d7d6234954ba",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Downloading...\n",
      "From: https://drive.google.com/uc?id=1iZ5Am1uS2keIdXqnsZh4Do6y1v_LlWGd\n",
      "To: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/mye.h5ad\n",
      "100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 43.2M/43.2M [00:00<00:00, 85.0MB/s]\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'/work/magroup/shared/Heimdall/data/cell_type_annotation/raw/mye.h5ad'"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gdown.download(\n",
    "    url=\"https://drive.google.com/file/d/1iZ5Am1uS2keIdXqnsZh4Do6y1v_LlWGd/view?usp=drive_link\",\n",
    "    output=str(RAW_DATA_DIR / \"mye.h5ad\"),\n",
    "    fuzzy=True,\n",
    ")\n",
    "\n",
    "# gdown.download(\n",
    "#     url=\"https://drive.google.com/file/d/1U556qLQMUCX1i2VZjJOVhPmpYSo71ZAj/view?usp=drive_link\",\n",
    "#     output=str(RAW_DATA_DIR / \"mye_ref.h5ad\"),\n",
    "#     fuzzy=True,\n",
    "# )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "37470bff-3195-4440-82ef-48345a387a68",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 3430 × 3000\n",
       "    obs: 'cell_type', 'cancer_type', 'batch'\n",
       "    uns: 'cancer_type_colors', 'cell_type_colors', 'log1p', 'neighbors', 'pca', 'umap'\n",
       "    obsm: 'X_pca', 'X_umap'\n",
       "    varm: 'PCs'\n",
       "    obsp: 'connectivities', 'distances'"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load raw data\n",
    "data = ad.read_h5ad(RAW_DATA_DIR / \"mye.h5ad\")\n",
    "data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "e3bf94ca-4146-4c7a-a8bd-98c4d5f40829",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 3430 × 2455\n",
       "    obs: 'cell_type', 'cancer_type', 'batch'\n",
       "    var: 'n_cells'\n",
       "    uns: 'cancer_type_colors', 'cell_type_colors', 'log1p', 'neighbors', 'pca', 'umap'\n",
       "    obsm: 'X_pca', 'X_umap'\n",
       "    varm: 'PCs'\n",
       "    obsp: 'connectivities', 'distances'"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Filtering\n",
    "sc.pp.filter_genes(data, min_cells=100, inplace=True)\n",
    "data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "a5932537-1e73-44d0-a984-856876b8392a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Mapping data directory: /work/magroup/shared/Heimdall/data/gene_mapping/ensembl/human\n",
      "Loading mapping from cache: /work/magroup/shared/Heimdall/data/gene_mapping/ensembl/human/symbol_to_ensembl.json\n",
      "Successfully mapped 2,267 out of 2,455 genes (92.3%)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>n_cells</th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>gene_ensembl</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000188290</th>\n",
       "      <td>811</td>\n",
       "      <td>HES4</td>\n",
       "      <td>ENSG00000188290</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000187608</th>\n",
       "      <td>1524</td>\n",
       "      <td>ISG15</td>\n",
       "      <td>ENSG00000187608</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000188157</th>\n",
       "      <td>103</td>\n",
       "      <td>AGRN</td>\n",
       "      <td>ENSG00000188157</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000186827</th>\n",
       "      <td>129</td>\n",
       "      <td>TNFRSF4</td>\n",
       "      <td>ENSG00000186827</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000248333</th>\n",
       "      <td>566</td>\n",
       "      <td>CDK11B</td>\n",
       "      <td>ENSG00000248333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000160285</th>\n",
       "      <td>127</td>\n",
       "      <td>LSS</td>\n",
       "      <td>ENSG00000160285</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000160307</th>\n",
       "      <td>191</td>\n",
       "      <td>S100B</td>\n",
       "      <td>ENSG00000160307</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000198899</th>\n",
       "      <td>3395</td>\n",
       "      <td>MT-ATP6</td>\n",
       "      <td>ENSG00000198899</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000198840</th>\n",
       "      <td>3325</td>\n",
       "      <td>MT-ND3</td>\n",
       "      <td>ENSG00000198840</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000198695</th>\n",
       "      <td>2176</td>\n",
       "      <td>MT-ND6</td>\n",
       "      <td>ENSG00000198695</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2455 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 n_cells gene_symbol     gene_ensembl\n",
       "ENSG00000188290      811        HES4  ENSG00000188290\n",
       "ENSG00000187608     1524       ISG15  ENSG00000187608\n",
       "ENSG00000188157      103        AGRN  ENSG00000188157\n",
       "ENSG00000186827      129     TNFRSF4  ENSG00000186827\n",
       "ENSG00000248333      566      CDK11B  ENSG00000248333\n",
       "...                  ...         ...              ...\n",
       "ENSG00000160285      127         LSS  ENSG00000160285\n",
       "ENSG00000160307      191       S100B  ENSG00000160307\n",
       "ENSG00000198899     3395     MT-ATP6  ENSG00000198899\n",
       "ENSG00000198840     3325      MT-ND3  ENSG00000198840\n",
       "ENSG00000198695     2176      MT-ND6  ENSG00000198695\n",
       "\n",
       "[2455 rows x 3 columns]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Gene id mapping\n",
    "symbol_to_ensembl_mapping = data_processing_utils.symbol_to_ensembl_from_ensembl(\n",
    "    data_dir=PROJECT_DIR / \"data\", genes=data.var.index.tolist(), species=\"human\")\n",
    "data.uns[\"gene_mapping:symbol_to_ensembl\"] = symbol_to_ensembl_mapping.mapping_full\n",
    "\n",
    "data.var[\"gene_symbol\"] = data.var.index\n",
    "data.var[\"gene_ensembl\"] = data.var[\"gene_symbol\"].map(symbol_to_ensembl_mapping.mapping_combined.get)\n",
    "data.var.index = data.var.index.map(symbol_to_ensembl_mapping.mapping_reduced)\n",
    "\n",
    "data.var"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "ad6467e7-6d8b-4ca6-a40b-bfae11444243",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 3430 × 2455\n",
       "    obs: 'celltype', 'cancer_type', 'batch', 'species', 'task_celltype'\n",
       "    var: 'n_cells', 'gene_symbol', 'gene_ensembl'\n",
       "    uns: 'cancer_type_colors', 'cell_type_colors', 'log1p', 'neighbors', 'pca', 'umap', 'gene_mapping:symbol_to_ensembl', 'celltype_order', 'batch_order'\n",
       "    obsm: 'X_pca', 'X_umap'\n",
       "    varm: 'PCs'\n",
       "    obsp: 'connectivities', 'distances'"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Standardize attributes\n",
    "data.obs[\"species\"] = \"Homo sapiens\"\n",
    "data.obs.rename(columns={\"cell_type\": \"celltype\"}, inplace=True)\n",
    "\n",
    "le = LabelEncoder()\n",
    "\n",
    "data.obs[\"task_celltype\"] = le.fit_transform(data.obs[\"celltype\"])\n",
    "data.uns[\"celltype_order\"] = le.classes_.tolist()\n",
    "\n",
    "data.obs[\"batch\"] = le.fit_transform(data.obs[\"batch\"])\n",
    "data.uns[\"batch_order\"] = le.classes_.tolist()\n",
    "\n",
    "data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "4d31fa62-8cc2-4d94-9dd1-4b2b10e298db",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save processed data\n",
    "data.write_h5ad(PROCESSED_DATA_DIR / \"mye.h5ad\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0f8c057d-0af5-4995-b31b-d5bc1e899a54",
   "metadata": {},
   "source": [
    "## Tabula Muris\n",
    "\n",
    "https://tabula-muris.ds.czbiohub.org/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "4f752b1b-deb7-4e99-8e67-0497756cf098",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--2024-07-10 11:08:00--  https://figshare.com/ndownloader/articles/5968960/versions/3\n",
      "Resolving figshare.com (figshare.com)... 52.215.106.69, 52.212.236.71, 2a05:d018:1f4:d000:f6c4:ca19:842d:9398, ...\n",
      "Connecting to figshare.com (figshare.com)|52.215.106.69|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 400440574 (382M) [application/zip]\n",
      "Saving to: ‘/work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm.zip’\n",
      "\n",
      "100%[======================================>] 400,440,574 28.1MB/s   in 14s    \n",
      "\n",
      "2024-07-10 11:08:15 (26.5 MB/s) - ‘/work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm.zip’ saved [400440574/400440574]\n",
      "\n",
      "Archive:  /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm.zip\n",
      " extracting: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/metadata_droplet.csv  \n",
      " extracting: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/droplet.zip  \n",
      " extracting: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/annotations_droplet.csv  \n",
      "Archive:  /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/droplet.zip\n",
      "   creating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/droplet/\n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/droplet/.DS_Store  \n",
      "   creating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/__MACOSX/\n",
      "   creating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/__MACOSX/droplet/\n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/__MACOSX/droplet/._.DS_Store  \n",
      "   creating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/droplet/Bladder-10X_P4_3/\n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/droplet/Bladder-10X_P4_3/barcodes.tsv  \n",
      "   creating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/__MACOSX/droplet/Bladder-10X_P4_3/\n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/__MACOSX/droplet/Bladder-10X_P4_3/._barcodes.tsv  \n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/droplet/Bladder-10X_P4_3/genes.tsv  \n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/__MACOSX/droplet/Bladder-10X_P4_3/._genes.tsv  \n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/droplet/Bladder-10X_P4_3/matrix.mtx  \n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/__MACOSX/droplet/Bladder-10X_P4_3/._matrix.mtx  \n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/__MACOSX/droplet/._Bladder-10X_P4_3  \n",
      "   creating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/droplet/Bladder-10X_P4_4/\n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/droplet/Bladder-10X_P4_4/barcodes.tsv  \n",
      "   creating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/__MACOSX/droplet/Bladder-10X_P4_4/\n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/__MACOSX/droplet/Bladder-10X_P4_4/._barcodes.tsv  \n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/droplet/Bladder-10X_P4_4/genes.tsv  \n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/__MACOSX/droplet/Bladder-10X_P4_4/._genes.tsv  \n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/droplet/Bladder-10X_P4_4/matrix.mtx  \n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/__MACOSX/droplet/Bladder-10X_P4_4/._matrix.mtx  \n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/__MACOSX/droplet/._Bladder-10X_P4_4  \n",
      "   creating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/droplet/Bladder-10X_P7_7/\n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/droplet/Bladder-10X_P7_7/barcodes.tsv  \n",
      "   creating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/__MACOSX/droplet/Bladder-10X_P7_7/\n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/__MACOSX/droplet/Bladder-10X_P7_7/._barcodes.tsv  \n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/droplet/Bladder-10X_P7_7/genes.tsv  \n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/__MACOSX/droplet/Bladder-10X_P7_7/._genes.tsv  \n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/droplet/Bladder-10X_P7_7/matrix.mtx  \n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/__MACOSX/droplet/Bladder-10X_P7_7/._matrix.mtx  \n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/__MACOSX/droplet/._Bladder-10X_P7_7  \n",
      "   creating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/droplet/Heart_and_Aorta-10X_P7_4/\n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/droplet/Heart_and_Aorta-10X_P7_4/barcodes.tsv  \n",
      "   creating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/__MACOSX/droplet/Heart_and_Aorta-10X_P7_4/\n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/__MACOSX/droplet/Heart_and_Aorta-10X_P7_4/._barcodes.tsv  \n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/droplet/Heart_and_Aorta-10X_P7_4/genes.tsv  \n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/__MACOSX/droplet/Heart_and_Aorta-10X_P7_4/._genes.tsv  \n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/droplet/Heart_and_Aorta-10X_P7_4/matrix.mtx  \n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/__MACOSX/droplet/Heart_and_Aorta-10X_P7_4/._matrix.mtx  \n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/__MACOSX/droplet/._Heart_and_Aorta-10X_P7_4  \n",
      "   creating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/droplet/Kidney-10X_P4_5/\n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/droplet/Kidney-10X_P4_5/barcodes.tsv  \n",
      "   creating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/__MACOSX/droplet/Kidney-10X_P4_5/\n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/__MACOSX/droplet/Kidney-10X_P4_5/._barcodes.tsv  \n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/droplet/Kidney-10X_P4_5/genes.tsv  \n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/__MACOSX/droplet/Kidney-10X_P4_5/._genes.tsv  \n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/droplet/Kidney-10X_P4_5/matrix.mtx  \n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/__MACOSX/droplet/Kidney-10X_P4_5/._matrix.mtx  \n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/__MACOSX/droplet/._Kidney-10X_P4_5  \n",
      "   creating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/droplet/Kidney-10X_P4_6/\n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/droplet/Kidney-10X_P4_6/barcodes.tsv  \n",
      "   creating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/__MACOSX/droplet/Kidney-10X_P4_6/\n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/__MACOSX/droplet/Kidney-10X_P4_6/._barcodes.tsv  \n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/droplet/Kidney-10X_P4_6/genes.tsv  \n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/__MACOSX/droplet/Kidney-10X_P4_6/._genes.tsv  \n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/droplet/Kidney-10X_P4_6/matrix.mtx  \n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/__MACOSX/droplet/Kidney-10X_P4_6/._matrix.mtx  \n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/__MACOSX/droplet/._Kidney-10X_P4_6  \n",
      "   creating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/droplet/Kidney-10X_P7_5/\n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/droplet/Kidney-10X_P7_5/barcodes.tsv  \n",
      "   creating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/__MACOSX/droplet/Kidney-10X_P7_5/\n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/__MACOSX/droplet/Kidney-10X_P7_5/._barcodes.tsv  \n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/droplet/Kidney-10X_P7_5/genes.tsv  \n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/__MACOSX/droplet/Kidney-10X_P7_5/._genes.tsv  \n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/droplet/Kidney-10X_P7_5/matrix.mtx  \n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/__MACOSX/droplet/Kidney-10X_P7_5/._matrix.mtx  \n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/__MACOSX/droplet/._Kidney-10X_P7_5  \n",
      "   creating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/droplet/Limb_Muscle-10X_P7_14/\n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/droplet/Limb_Muscle-10X_P7_14/barcodes.tsv  \n",
      "   creating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/__MACOSX/droplet/Limb_Muscle-10X_P7_14/\n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/__MACOSX/droplet/Limb_Muscle-10X_P7_14/._barcodes.tsv  \n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/droplet/Limb_Muscle-10X_P7_14/genes.tsv  \n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/__MACOSX/droplet/Limb_Muscle-10X_P7_14/._genes.tsv  \n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/droplet/Limb_Muscle-10X_P7_14/matrix.mtx  \n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/__MACOSX/droplet/Limb_Muscle-10X_P7_14/._matrix.mtx  \n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/__MACOSX/droplet/._Limb_Muscle-10X_P7_14  \n",
      "   creating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/droplet/Limb_Muscle-10X_P7_15/\n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/droplet/Limb_Muscle-10X_P7_15/barcodes.tsv  \n",
      "   creating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/__MACOSX/droplet/Limb_Muscle-10X_P7_15/\n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/__MACOSX/droplet/Limb_Muscle-10X_P7_15/._barcodes.tsv  \n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/droplet/Limb_Muscle-10X_P7_15/genes.tsv  \n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/__MACOSX/droplet/Limb_Muscle-10X_P7_15/._genes.tsv  \n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/droplet/Limb_Muscle-10X_P7_15/matrix.mtx  \n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/__MACOSX/droplet/Limb_Muscle-10X_P7_15/._matrix.mtx  \n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/__MACOSX/droplet/._Limb_Muscle-10X_P7_15  \n",
      "   creating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/droplet/Liver-10X_P4_2/\n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/droplet/Liver-10X_P4_2/barcodes.tsv  \n",
      "   creating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/__MACOSX/droplet/Liver-10X_P4_2/\n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/__MACOSX/droplet/Liver-10X_P4_2/._barcodes.tsv  \n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/droplet/Liver-10X_P4_2/genes.tsv  \n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/__MACOSX/droplet/Liver-10X_P4_2/._genes.tsv  \n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/droplet/Liver-10X_P4_2/matrix.mtx  \n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/__MACOSX/droplet/Liver-10X_P4_2/._matrix.mtx  \n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/__MACOSX/droplet/._Liver-10X_P4_2  \n",
      "   creating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/droplet/Liver-10X_P7_0/\n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/droplet/Liver-10X_P7_0/barcodes.tsv  \n",
      "   creating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/__MACOSX/droplet/Liver-10X_P7_0/\n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/__MACOSX/droplet/Liver-10X_P7_0/._barcodes.tsv  \n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/droplet/Liver-10X_P7_0/genes.tsv  \n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/__MACOSX/droplet/Liver-10X_P7_0/._genes.tsv  \n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/droplet/Liver-10X_P7_0/matrix.mtx  \n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/__MACOSX/droplet/Liver-10X_P7_0/._matrix.mtx  \n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/__MACOSX/droplet/._Liver-10X_P7_0  \n",
      "   creating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/droplet/Liver-10X_P7_1/\n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/droplet/Liver-10X_P7_1/barcodes.tsv  \n",
      "   creating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/__MACOSX/droplet/Liver-10X_P7_1/\n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/__MACOSX/droplet/Liver-10X_P7_1/._barcodes.tsv  \n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/droplet/Liver-10X_P7_1/genes.tsv  \n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/__MACOSX/droplet/Liver-10X_P7_1/._genes.tsv  \n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/droplet/Liver-10X_P7_1/matrix.mtx  \n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/__MACOSX/droplet/Liver-10X_P7_1/._matrix.mtx  \n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/__MACOSX/droplet/._Liver-10X_P7_1  \n",
      "   creating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/droplet/Lung-10X_P7_8/\n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/droplet/Lung-10X_P7_8/barcodes.tsv  \n",
      "   creating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/__MACOSX/droplet/Lung-10X_P7_8/\n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/__MACOSX/droplet/Lung-10X_P7_8/._barcodes.tsv  \n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/droplet/Lung-10X_P7_8/genes.tsv  \n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/__MACOSX/droplet/Lung-10X_P7_8/._genes.tsv  \n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/droplet/Lung-10X_P7_8/matrix.mtx  \n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/__MACOSX/droplet/Lung-10X_P7_8/._matrix.mtx  \n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/__MACOSX/droplet/._Lung-10X_P7_8  \n",
      "   creating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/droplet/Lung-10X_P7_9/\n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/droplet/Lung-10X_P7_9/barcodes.tsv  \n",
      "   creating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/__MACOSX/droplet/Lung-10X_P7_9/\n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/__MACOSX/droplet/Lung-10X_P7_9/._barcodes.tsv  \n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/droplet/Lung-10X_P7_9/genes.tsv  \n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/__MACOSX/droplet/Lung-10X_P7_9/._genes.tsv  \n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/droplet/Lung-10X_P7_9/matrix.mtx  \n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/__MACOSX/droplet/Lung-10X_P7_9/._matrix.mtx  \n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/__MACOSX/droplet/._Lung-10X_P7_9  \n",
      "   creating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/droplet/Lung-10X_P8_12/\n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/droplet/Lung-10X_P8_12/barcodes.tsv  \n",
      "   creating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/__MACOSX/droplet/Lung-10X_P8_12/\n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/__MACOSX/droplet/Lung-10X_P8_12/._barcodes.tsv  \n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/droplet/Lung-10X_P8_12/genes.tsv  \n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/__MACOSX/droplet/Lung-10X_P8_12/._genes.tsv  \n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/droplet/Lung-10X_P8_12/matrix.mtx  \n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/__MACOSX/droplet/Lung-10X_P8_12/._matrix.mtx  \n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/__MACOSX/droplet/._Lung-10X_P8_12  \n",
      "   creating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/droplet/Lung-10X_P8_13/\n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/droplet/Lung-10X_P8_13/barcodes.tsv  \n",
      "   creating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/__MACOSX/droplet/Lung-10X_P8_13/\n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/__MACOSX/droplet/Lung-10X_P8_13/._barcodes.tsv  \n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/droplet/Lung-10X_P8_13/genes.tsv  \n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/__MACOSX/droplet/Lung-10X_P8_13/._genes.tsv  \n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/droplet/Lung-10X_P8_13/matrix.mtx  \n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/__MACOSX/droplet/Lung-10X_P8_13/._matrix.mtx  \n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/__MACOSX/droplet/._Lung-10X_P8_13  \n",
      "   creating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/droplet/Mammary_Gland-10X_P7_12/\n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/droplet/Mammary_Gland-10X_P7_12/barcodes.tsv  \n",
      "   creating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/__MACOSX/droplet/Mammary_Gland-10X_P7_12/\n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/__MACOSX/droplet/Mammary_Gland-10X_P7_12/._barcodes.tsv  \n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/droplet/Mammary_Gland-10X_P7_12/genes.tsv  \n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/__MACOSX/droplet/Mammary_Gland-10X_P7_12/._genes.tsv  \n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/droplet/Mammary_Gland-10X_P7_12/matrix.mtx  \n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/__MACOSX/droplet/Mammary_Gland-10X_P7_12/._matrix.mtx  \n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/__MACOSX/droplet/._Mammary_Gland-10X_P7_12  \n",
      "   creating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/droplet/Mammary_Gland-10X_P7_13/\n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/droplet/Mammary_Gland-10X_P7_13/barcodes.tsv  \n",
      "   creating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/__MACOSX/droplet/Mammary_Gland-10X_P7_13/\n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/__MACOSX/droplet/Mammary_Gland-10X_P7_13/._barcodes.tsv  \n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/droplet/Mammary_Gland-10X_P7_13/genes.tsv  \n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/__MACOSX/droplet/Mammary_Gland-10X_P7_13/._genes.tsv  \n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/droplet/Mammary_Gland-10X_P7_13/matrix.mtx  \n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/__MACOSX/droplet/Mammary_Gland-10X_P7_13/._matrix.mtx  \n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/__MACOSX/droplet/._Mammary_Gland-10X_P7_13  \n",
      "   creating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/droplet/Marrow-10X_P7_2/\n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/droplet/Marrow-10X_P7_2/barcodes.tsv  \n",
      "   creating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/__MACOSX/droplet/Marrow-10X_P7_2/\n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/__MACOSX/droplet/Marrow-10X_P7_2/._barcodes.tsv  \n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/droplet/Marrow-10X_P7_2/genes.tsv  \n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/__MACOSX/droplet/Marrow-10X_P7_2/._genes.tsv  \n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/droplet/Marrow-10X_P7_2/matrix.mtx  \n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/__MACOSX/droplet/Marrow-10X_P7_2/._matrix.mtx  \n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/__MACOSX/droplet/._Marrow-10X_P7_2  \n",
      "   creating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/droplet/Marrow-10X_P7_3/\n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/droplet/Marrow-10X_P7_3/barcodes.tsv  \n",
      "   creating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/__MACOSX/droplet/Marrow-10X_P7_3/\n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/__MACOSX/droplet/Marrow-10X_P7_3/._barcodes.tsv  \n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/droplet/Marrow-10X_P7_3/genes.tsv  \n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/__MACOSX/droplet/Marrow-10X_P7_3/._genes.tsv  \n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/droplet/Marrow-10X_P7_3/matrix.mtx  \n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/__MACOSX/droplet/Marrow-10X_P7_3/._matrix.mtx  \n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/__MACOSX/droplet/._Marrow-10X_P7_3  \n",
      "   creating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/droplet/Spleen-10X_P4_7/\n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/droplet/Spleen-10X_P4_7/barcodes.tsv  \n",
      "   creating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/__MACOSX/droplet/Spleen-10X_P4_7/\n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/__MACOSX/droplet/Spleen-10X_P4_7/._barcodes.tsv  \n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/droplet/Spleen-10X_P4_7/genes.tsv  \n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/__MACOSX/droplet/Spleen-10X_P4_7/._genes.tsv  \n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/droplet/Spleen-10X_P4_7/matrix.mtx  \n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/__MACOSX/droplet/Spleen-10X_P4_7/._matrix.mtx  \n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/__MACOSX/droplet/._Spleen-10X_P4_7  \n",
      "   creating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/droplet/Spleen-10X_P7_6/\n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/droplet/Spleen-10X_P7_6/barcodes.tsv  \n",
      "   creating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/__MACOSX/droplet/Spleen-10X_P7_6/\n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/__MACOSX/droplet/Spleen-10X_P7_6/._barcodes.tsv  \n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/droplet/Spleen-10X_P7_6/genes.tsv  \n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/__MACOSX/droplet/Spleen-10X_P7_6/._genes.tsv  \n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/droplet/Spleen-10X_P7_6/matrix.mtx  \n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/__MACOSX/droplet/Spleen-10X_P7_6/._matrix.mtx  \n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/__MACOSX/droplet/._Spleen-10X_P7_6  \n",
      "   creating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/droplet/Thymus-10X_P7_11/\n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/droplet/Thymus-10X_P7_11/barcodes.tsv  \n",
      "   creating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/__MACOSX/droplet/Thymus-10X_P7_11/\n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/__MACOSX/droplet/Thymus-10X_P7_11/._barcodes.tsv  \n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/droplet/Thymus-10X_P7_11/genes.tsv  \n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/__MACOSX/droplet/Thymus-10X_P7_11/._genes.tsv  \n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/droplet/Thymus-10X_P7_11/matrix.mtx  \n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/__MACOSX/droplet/Thymus-10X_P7_11/._matrix.mtx  \n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/__MACOSX/droplet/._Thymus-10X_P7_11  \n",
      "   creating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/droplet/Tongue-10X_P4_0/\n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/droplet/Tongue-10X_P4_0/barcodes.tsv  \n",
      "   creating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/__MACOSX/droplet/Tongue-10X_P4_0/\n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/__MACOSX/droplet/Tongue-10X_P4_0/._barcodes.tsv  \n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/droplet/Tongue-10X_P4_0/genes.tsv  \n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/__MACOSX/droplet/Tongue-10X_P4_0/._genes.tsv  \n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/droplet/Tongue-10X_P4_0/matrix.mtx  \n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/__MACOSX/droplet/Tongue-10X_P4_0/._matrix.mtx  \n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/__MACOSX/droplet/._Tongue-10X_P4_0  \n",
      "   creating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/droplet/Tongue-10X_P4_1/\n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/droplet/Tongue-10X_P4_1/barcodes.tsv  \n",
      "   creating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/__MACOSX/droplet/Tongue-10X_P4_1/\n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/__MACOSX/droplet/Tongue-10X_P4_1/._barcodes.tsv  \n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/droplet/Tongue-10X_P4_1/genes.tsv  \n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/__MACOSX/droplet/Tongue-10X_P4_1/._genes.tsv  \n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/droplet/Tongue-10X_P4_1/matrix.mtx  \n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/__MACOSX/droplet/Tongue-10X_P4_1/._matrix.mtx  \n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/__MACOSX/droplet/._Tongue-10X_P4_1  \n",
      "   creating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/droplet/Tongue-10X_P7_10/\n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/droplet/Tongue-10X_P7_10/barcodes.tsv  \n",
      "   creating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/__MACOSX/droplet/Tongue-10X_P7_10/\n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/__MACOSX/droplet/Tongue-10X_P7_10/._barcodes.tsv  \n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/droplet/Tongue-10X_P7_10/genes.tsv  \n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/__MACOSX/droplet/Tongue-10X_P7_10/._genes.tsv  \n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/droplet/Tongue-10X_P7_10/matrix.mtx  \n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/__MACOSX/droplet/Tongue-10X_P7_10/._matrix.mtx  \n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/__MACOSX/droplet/._Tongue-10X_P7_10  \n",
      "   creating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/droplet/Trachea-10X_P8_14/\n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/droplet/Trachea-10X_P8_14/barcodes.tsv  \n",
      "   creating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/__MACOSX/droplet/Trachea-10X_P8_14/\n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/__MACOSX/droplet/Trachea-10X_P8_14/._barcodes.tsv  \n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/droplet/Trachea-10X_P8_14/genes.tsv  \n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/__MACOSX/droplet/Trachea-10X_P8_14/._genes.tsv  \n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/droplet/Trachea-10X_P8_14/matrix.mtx  \n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/__MACOSX/droplet/Trachea-10X_P8_14/._matrix.mtx  \n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/__MACOSX/droplet/._Trachea-10X_P8_14  \n",
      "   creating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/droplet/Trachea-10X_P8_15/\n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/droplet/Trachea-10X_P8_15/barcodes.tsv  \n",
      "   creating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/__MACOSX/droplet/Trachea-10X_P8_15/\n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/__MACOSX/droplet/Trachea-10X_P8_15/._barcodes.tsv  \n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/droplet/Trachea-10X_P8_15/genes.tsv  \n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/__MACOSX/droplet/Trachea-10X_P8_15/._genes.tsv  \n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/droplet/Trachea-10X_P8_15/matrix.mtx  \n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/__MACOSX/droplet/Trachea-10X_P8_15/._matrix.mtx  \n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/__MACOSX/droplet/._Trachea-10X_P8_15  \n",
      "  inflating: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw/tm/__MACOSX/._droplet  \n"
     ]
    }
   ],
   "source": [
    "!wget https://figshare.com/ndownloader/articles/5968960/versions/3 -O {RAW_DATA_DIR}/tm.zip\n",
    "!unzip {RAW_DATA_DIR}/tm.zip -d {RAW_DATA_DIR}/tm\n",
    "!unzip {RAW_DATA_DIR}/tm/droplet.zip -d {RAW_DATA_DIR}/tm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "4f037b30-058e-4ab8-b35b-b8214f958a25",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Extracting Marrow-10X_P7_3\n",
      "Extracting Kidney-10X_P7_5\n",
      "Extracting Kidney-10X_P4_6\n",
      "Extracting Lung-10X_P7_9\n",
      "Extracting Spleen-10X_P4_7\n",
      "Extracting Lung-10X_P8_13\n",
      "Extracting Kidney-10X_P4_5\n",
      "Extracting Mammary_Gland-10X_P7_13\n",
      "Extracting Tongue-10X_P4_1\n",
      "Extracting Trachea-10X_P8_15\n",
      "Extracting Liver-10X_P7_0\n",
      "Extracting Limb_Muscle-10X_P7_15\n",
      "Extracting Heart_and_Aorta-10X_P7_4\n",
      "Extracting Bladder-10X_P4_3\n",
      "Extracting Lung-10X_P7_8\n",
      "Extracting Bladder-10X_P4_4\n",
      "Extracting Bladder-10X_P7_7\n",
      "Extracting Marrow-10X_P7_2\n",
      "Extracting Tongue-10X_P7_10\n",
      "Extracting Limb_Muscle-10X_P7_14\n",
      "Extracting Thymus-10X_P7_11\n",
      "Extracting Spleen-10X_P7_6\n",
      "Extracting Tongue-10X_P4_0\n",
      "Extracting Mammary_Gland-10X_P7_12\n",
      "Extracting Trachea-10X_P8_14\n",
      "Extracting Liver-10X_P4_2\n",
      "Extracting Liver-10X_P7_1\n",
      "Extracting Lung-10X_P8_12\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 55656 × 23433\n",
       "    obs: 'cell_ontology_class', 'cell_ontology_id', 'channel', 'cluster.ids', 'mouse.id', 'mouse.sex', 'tissue', 'tissue_tSNE_1', 'tissue_tSNE_2'"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load raw data\n",
    "data_list = []\n",
    "\n",
    "tm_data_dir = RAW_DATA_DIR / \"tm\"\n",
    "tm_droplet_data_dir = tm_data_dir / \"droplet\"\n",
    "\n",
    "annot_df = pd.read_csv(tm_data_dir / \"annotations_droplet.csv\", low_memory=False, index_col=0)\n",
    "annot_df.index.name = None\n",
    "annot_df = annot_df[[\n",
    " \"cell_ontology_class\",\n",
    " \"cell_ontology_id\",\n",
    " \"channel\",\n",
    " \"cluster.ids\",\n",
    " \"mouse.id\",\n",
    " \"mouse.sex\",\n",
    " \"tissue\",\n",
    " \"tissue_tSNE_1\",\n",
    " \"tissue_tSNE_2\",\n",
    "]].copy()  # exclude object typed (those with NaN's) columns (https://github.com/scverse/anndata/issues/558)\n",
    "\n",
    "all_cells = set(annot_df.index)\n",
    "\n",
    "for i in os.listdir(tm_droplet_data_dir):\n",
    "    if i == \".DS_Store\":\n",
    "        continue\n",
    "\n",
    "    print(f\"Extracting {i}\")\n",
    "    tissue, channel = i.split(\"-\")\n",
    "\n",
    "    bcs = pd.read_csv(tm_droplet_data_dir / i / \"barcodes.tsv\", sep=\"\\t\", header=None, names=[\"raw_cell_id\"])\n",
    "    genes = pd.read_csv(tm_droplet_data_dir / i / \"genes.tsv\", sep=\"\\t\", header=None)\n",
    "    data_partial = ad.read_mtx(tm_droplet_data_dir / i / \"matrix.mtx\").T\n",
    "\n",
    "    bcs[\"cell_id\"] = [\"_\".join((channel, i.split(\"-\")[0])) for i in bcs[\"raw_cell_id\"]]\n",
    "    valid_cells = [i for i in bcs[\"cell_id\"] if  i in all_cells]\n",
    "    # print(f\"Matched {len(valid_cells):,} out of {len(bcs):,} cells\")\n",
    "\n",
    "    bcs.set_index(\"cell_id\", inplace=True)\n",
    "    bcs.index.name = None\n",
    "\n",
    "    genes.set_index(0, inplace=True)\n",
    "    genes.index.name = None\n",
    "\n",
    "    data_partial.var = genes\n",
    "    data_partial.obs = bcs\n",
    "    data_list.append(data_partial[valid_cells].copy())\n",
    "\n",
    "data = ad.concat(data_list)[annot_df.index].copy()\n",
    "data.obs = annot_df\n",
    "\n",
    "data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "3085317e-dbda-404a-8ee7-820798b5ffc0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 55656 × 15815\n",
       "    obs: 'cell_ontology_class', 'cell_ontology_id', 'channel', 'cluster.ids', 'mouse.id', 'mouse.sex', 'tissue', 'tissue_tSNE_1', 'tissue_tSNE_2'\n",
       "    var: 'n_cells'"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Filtering\n",
    "sc.pp.filter_genes(data, min_cells=100, inplace=True)\n",
    "data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "a2a7266d-aa30-4f18-b992-e60e167b1daf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 55656 × 15815\n",
       "    obs: 'cell_ontology_class', 'cell_ontology_id', 'channel', 'cluster.ids', 'mouse.id', 'mouse.sex', 'tissue', 'tissue_tSNE_1', 'tissue_tSNE_2'\n",
       "    var: 'n_cells'\n",
       "    uns: 'log1p'"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Normalize\n",
    "sc.pp.normalize_total(data, target_sum=10000)\n",
    "sc.pp.log1p(data)\n",
    "data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "72599360-232e-4c92-977e-35d7dae4adbf",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Mapping data directory: /work/magroup/shared/Heimdall/data/gene_mapping/ensembl/mouse\n",
      "Loading mapping from cache: /work/magroup/shared/Heimdall/data/gene_mapping/ensembl/mouse/symbol_to_ensembl.json\n",
      "Successfully mapped 13,867 out of 15,815 genes (87.7%)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>n_cells</th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>gene_ensembl</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000025902</th>\n",
       "      <td>2049</td>\n",
       "      <td>Sox17</td>\n",
       "      <td>ENSMUSG00000025902</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000033845</th>\n",
       "      <td>16312</td>\n",
       "      <td>Mrpl15</td>\n",
       "      <td>ENSMUSG00000033845</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000025903</th>\n",
       "      <td>15290</td>\n",
       "      <td>Lypla1</td>\n",
       "      <td>ENSMUSG00000025903</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000033813</th>\n",
       "      <td>21414</td>\n",
       "      <td>Tcea1</td>\n",
       "      <td>ENSMUSG00000033813</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000002459</th>\n",
       "      <td>392</td>\n",
       "      <td>Rgs20</td>\n",
       "      <td>ENSMUSG00000002459</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000056673</th>\n",
       "      <td>1086</td>\n",
       "      <td>Kdm5d</td>\n",
       "      <td>ENSMUSG00000056673</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000069049</th>\n",
       "      <td>7841</td>\n",
       "      <td>Eif2s3y</td>\n",
       "      <td>ENSMUSG00000069049</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000068457</th>\n",
       "      <td>851</td>\n",
       "      <td>Uty</td>\n",
       "      <td>ENSMUSG00000068457</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000069045</th>\n",
       "      <td>10450</td>\n",
       "      <td>Ddx3y</td>\n",
       "      <td>ENSMUSG00000069045</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Erdr1</th>\n",
       "      <td>15929</td>\n",
       "      <td>Erdr1</td>\n",
       "      <td>N/A</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>15815 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    n_cells gene_symbol        gene_ensembl\n",
       "ENSMUSG00000025902     2049       Sox17  ENSMUSG00000025902\n",
       "ENSMUSG00000033845    16312      Mrpl15  ENSMUSG00000033845\n",
       "ENSMUSG00000025903    15290      Lypla1  ENSMUSG00000025903\n",
       "ENSMUSG00000033813    21414       Tcea1  ENSMUSG00000033813\n",
       "ENSMUSG00000002459      392       Rgs20  ENSMUSG00000002459\n",
       "...                     ...         ...                 ...\n",
       "ENSMUSG00000056673     1086       Kdm5d  ENSMUSG00000056673\n",
       "ENSMUSG00000069049     7841     Eif2s3y  ENSMUSG00000069049\n",
       "ENSMUSG00000068457      851         Uty  ENSMUSG00000068457\n",
       "ENSMUSG00000069045    10450       Ddx3y  ENSMUSG00000069045\n",
       "Erdr1                 15929       Erdr1                 N/A\n",
       "\n",
       "[15815 rows x 3 columns]"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Gene id mapping\n",
    "symbol_to_ensembl_mapping = data_processing_utils.symbol_to_ensembl_from_ensembl(\n",
    "    data_dir=PROJECT_DIR / \"data\", genes=data.var.index.tolist(), species=\"mouse\")\n",
    "data.uns[\"gene_mapping:symbol_to_ensembl\"] = symbol_to_ensembl_mapping.mapping_full\n",
    "\n",
    "data.var[\"gene_symbol\"] = data.var.index\n",
    "data.var[\"gene_ensembl\"] = data.var[\"gene_symbol\"].map(symbol_to_ensembl_mapping.mapping_combined.get)\n",
    "data.var.index = data.var.index.map(symbol_to_ensembl_mapping.mapping_reduced)\n",
    "\n",
    "data.var"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "2503a5c7-3c4b-4cf6-87a9-43b50666f1fb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 55656 × 15815\n",
       "    obs: 'celltype', 'cell_ontology_id', 'channel', 'cluster.ids', 'mouse.id', 'mouse.sex', 'tissue', 'tissue_tSNE_1', 'tissue_tSNE_2', 'species', 'task_celltype', 'batch'\n",
       "    var: 'n_cells', 'gene_symbol', 'gene_ensembl'\n",
       "    uns: 'log1p', 'gene_mapping:symbol_to_ensembl', 'celltype_order', 'batch_order'"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Standardize attributes\n",
    "data.obs[\"species\"] = \"Mus musculus\"\n",
    "data.obs.rename(columns={\"cell_ontology_class\": \"celltype\"}, inplace=True)\n",
    "\n",
    "le = LabelEncoder()\n",
    "\n",
    "data.obs[\"task_celltype\"] = le.fit_transform(data.obs[\"celltype\"])\n",
    "data.uns[\"celltype_order\"] = le.classes_.tolist()\n",
    "\n",
    "data.obs[\"batch\"] = le.fit_transform(data.obs[\"channel\"])\n",
    "data.uns[\"batch_order\"] = le.classes_.tolist()\n",
    "\n",
    "data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "6813f057-6f57-416d-a696-00e7d6eca848",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save processed data\n",
    "data.write_h5ad(PROCESSED_DATA_DIR / \"tm.h5ad\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4ea2574e-f2c9-4ba3-90df-e9c94bf4caf7",
   "metadata": {},
   "source": [
    "## Allen Mouse Brain (AMB)\n",
    "\n",
    "https://cellxgene.cziscience.com/collections/45f0f67d-4b69-4a3c-a4e8-a63b962e843f"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "dcb81d69-853c-4d57-b517-d409397c782f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--2024-07-10 11:10:56--  https://datasets.cellxgene.cziscience.com/1de08aa2-b3c6-4f2d-9429-5d0a6b716b5c.h5ad\n",
      "Resolving datasets.cellxgene.cziscience.com (datasets.cellxgene.cziscience.com)... 18.160.200.37, 18.160.200.32, 18.160.200.93, ...\n",
      "Connecting to datasets.cellxgene.cziscience.com (datasets.cellxgene.cziscience.com)|18.160.200.37|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 1298779249 (1.2G) [binary/octet-stream]\n",
      "Saving to: ‘/work/magroup/shared/Heimdall/data/cell_type_annotation/raw/amb.h5ad’\n",
      "\n",
      "100%[====================================>] 1,298,779,249  105MB/s   in 12s    \n",
      "\n",
      "2024-07-10 11:11:08 (106 MB/s) - ‘/work/magroup/shared/Heimdall/data/cell_type_annotation/raw/amb.h5ad’ saved [1298779249/1298779249]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "!wget https://datasets.cellxgene.cziscience.com/1de08aa2-b3c6-4f2d-9429-5d0a6b716b5c.h5ad -O {RAW_DATA_DIR}/amb.h5ad"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "274a4135-7529-4b25-9a1c-e1d618ef4e68",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 22375 × 35304\n",
       "    obs: 'donor_id', 'age_days', 'eye_condition', 'genotype', 'driver_lines', 'reporter_lines', 'brain_hemisphere', 'brain_region', 'brain_subregion', 'injection_label_direction', 'injection_primary', 'injection_secondary', 'injection_tract', 'injection_material', 'injection_exclusion_criterion', 'facs_date', 'facs_container', 'facs_sort_criteria', 'rna_amplification_set', 'rna_amplification_pcr_cycles', 'library_prep_set', 'library_prep_avg_size_bp', 'seq_tube', 'seq_batch', 'total_reads', 'percent_exon_reads', 'percent_intron_reads', 'percent_intergenic_reads', 'percent_rrna_reads', 'percent_mt_exon_reads', 'percent_reads_unique', 'percent_synth_reads', 'percent_ecoli_reads', 'percent_aligned_reads_total', 'complexity_cg', 'genes_detected_cpm_criterion', 'genes_detected_fpkm_criterion', 'tdt_cpm', 'gfp_cpm', 'BICCN_class_label', 'BICCN_subclass_label', 'cluster', 'confusion_score', 'cluster_correlation', 'core_intermediate_call', 'BICCN_ontology_term_id', 'assay_ontology_term_id', 'disease_ontology_term_id', 'tissue_ontology_term_id', 'development_stage_ontology_term_id', 'cell_type_ontology_term_id', 'self_reported_ethnicity_ontology_term_id', 'sex_ontology_term_id', 'is_primary_data', 'organism_ontology_term_id', 'suspension_type', 'tissue_type', 'cell_type', 'assay', 'disease', 'organism', 'sex', 'tissue', 'self_reported_ethnicity', 'development_stage', 'observation_joinid'\n",
       "    var: 'feature_is_filtered', 'feature_name', 'feature_reference', 'feature_biotype', 'feature_length'\n",
       "    uns: 'citation', 'schema_reference', 'schema_version', 'title'\n",
       "    obsm: 'X_tsne'"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load raw data\n",
    "data = ad.read_h5ad(RAW_DATA_DIR / \"amb.h5ad\")\n",
    "data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "681bdafa-ad2c-4e3a-8b1b-bee646df2dcb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 22375 × 23326\n",
       "    obs: 'donor_id', 'age_days', 'eye_condition', 'genotype', 'driver_lines', 'reporter_lines', 'brain_hemisphere', 'brain_region', 'brain_subregion', 'injection_label_direction', 'injection_primary', 'injection_secondary', 'injection_tract', 'injection_material', 'injection_exclusion_criterion', 'facs_date', 'facs_container', 'facs_sort_criteria', 'rna_amplification_set', 'rna_amplification_pcr_cycles', 'library_prep_set', 'library_prep_avg_size_bp', 'seq_tube', 'seq_batch', 'total_reads', 'percent_exon_reads', 'percent_intron_reads', 'percent_intergenic_reads', 'percent_rrna_reads', 'percent_mt_exon_reads', 'percent_reads_unique', 'percent_synth_reads', 'percent_ecoli_reads', 'percent_aligned_reads_total', 'complexity_cg', 'genes_detected_cpm_criterion', 'genes_detected_fpkm_criterion', 'tdt_cpm', 'gfp_cpm', 'BICCN_class_label', 'BICCN_subclass_label', 'cluster', 'confusion_score', 'cluster_correlation', 'core_intermediate_call', 'BICCN_ontology_term_id', 'assay_ontology_term_id', 'disease_ontology_term_id', 'tissue_ontology_term_id', 'development_stage_ontology_term_id', 'cell_type_ontology_term_id', 'self_reported_ethnicity_ontology_term_id', 'sex_ontology_term_id', 'is_primary_data', 'organism_ontology_term_id', 'suspension_type', 'tissue_type', 'cell_type', 'assay', 'disease', 'organism', 'sex', 'tissue', 'self_reported_ethnicity', 'development_stage', 'observation_joinid'\n",
       "    var: 'feature_is_filtered', 'feature_name', 'feature_reference', 'feature_biotype', 'feature_length', 'n_cells'\n",
       "    uns: 'citation', 'schema_reference', 'schema_version', 'title'\n",
       "    obsm: 'X_tsne'"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Filtering\n",
    "sc.pp.filter_genes(data, min_cells=100, inplace=True)\n",
    "data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "b37129cc-ded0-48f2-8853-36acc78a3a6d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 22375 × 23326\n",
       "    obs: 'donor_id', 'age_days', 'eye_condition', 'genotype', 'driver_lines', 'reporter_lines', 'brain_hemisphere', 'brain_region', 'brain_subregion', 'injection_label_direction', 'injection_primary', 'injection_secondary', 'injection_tract', 'injection_material', 'injection_exclusion_criterion', 'facs_date', 'facs_container', 'facs_sort_criteria', 'rna_amplification_set', 'rna_amplification_pcr_cycles', 'library_prep_set', 'library_prep_avg_size_bp', 'seq_tube', 'seq_batch', 'total_reads', 'percent_exon_reads', 'percent_intron_reads', 'percent_intergenic_reads', 'percent_rrna_reads', 'percent_mt_exon_reads', 'percent_reads_unique', 'percent_synth_reads', 'percent_ecoli_reads', 'percent_aligned_reads_total', 'complexity_cg', 'genes_detected_cpm_criterion', 'genes_detected_fpkm_criterion', 'tdt_cpm', 'gfp_cpm', 'BICCN_class_label', 'celltype', 'cluster', 'confusion_score', 'cluster_correlation', 'core_intermediate_call', 'BICCN_ontology_term_id', 'assay_ontology_term_id', 'disease_ontology_term_id', 'tissue_ontology_term_id', 'development_stage_ontology_term_id', 'cell_type_ontology_term_id', 'self_reported_ethnicity_ontology_term_id', 'sex_ontology_term_id', 'is_primary_data', 'organism_ontology_term_id', 'suspension_type', 'tissue_type', 'cell_type', 'assay', 'disease', 'species', 'sex', 'tissue', 'self_reported_ethnicity', 'development_stage', 'observation_joinid', 'task_celltype'\n",
       "    var: 'feature_is_filtered', 'feature_name', 'feature_reference', 'feature_biotype', 'feature_length', 'n_cells', 'gene_symbol', 'gene_ensembl'\n",
       "    uns: 'citation', 'schema_reference', 'schema_version', 'title', 'celltype_order'\n",
       "    obsm: 'X_tsne'"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Standardize attributes\n",
    "data.obs.rename(columns={\"BICCN_subclass_label\": \"celltype\", \"organism\": \"species\"}, inplace=True)\n",
    "data.var[\"gene_symbol\"] = data.var[\"feature_name\"]\n",
    "data.var[\"gene_ensembl\"] = data.var.index\n",
    "\n",
    "le = LabelEncoder()\n",
    "\n",
    "data.obs[\"task_celltype\"] = le.fit_transform(data.obs[\"celltype\"])\n",
    "data.uns[\"celltype_order\"] = le.classes_.tolist()\n",
    "\n",
    "data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "180df8b2-1dd3-4dc8-a1eb-aa4e396fe6e5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save processed data\n",
    "data.write_h5ad(PROCESSED_DATA_DIR / \"amb.h5ad\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2770f0f6-68b0-455a-9c65-d03fe12e19bb",
   "metadata": {},
   "source": [
    "## Mouse Pancreatic Islet (MPI)\n",
    "\n",
    "https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE211799"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "0560a5b7-67ac-44b1-88e4-34bf82582163",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--2024-07-10 11:12:25--  https://ftp.ncbi.nlm.nih.gov/geo/series/GSE211nnn/GSE211799/suppl/GSE211799%5Fadata%5Fatlas.h5ad.gz\n",
      "Resolving ftp.ncbi.nlm.nih.gov (ftp.ncbi.nlm.nih.gov)... 130.14.250.13, 130.14.250.12, 2607:f220:41e:250::12, ...\n",
      "Connecting to ftp.ncbi.nlm.nih.gov (ftp.ncbi.nlm.nih.gov)|130.14.250.13|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 4257833437 (4.0G) [application/x-gzip]\n",
      "Saving to: ‘/work/magroup/shared/Heimdall/data/cell_type_annotation/raw/mpi.h5ad.gz’\n",
      "\n",
      "100%[====================================>] 4,257,833,437 76.3MB/s   in 53s    \n",
      "\n",
      "2024-07-10 11:13:18 (76.7 MB/s) - ‘/work/magroup/shared/Heimdall/data/cell_type_annotation/raw/mpi.h5ad.gz’ saved [4257833437/4257833437]\n",
      "\n",
      "gzip: /work/magroup/shared/Heimdall/data/cell_type_annotation/raw is a directory -- ignored\n"
     ]
    }
   ],
   "source": [
    "!wget https://ftp.ncbi.nlm.nih.gov/geo/series/GSE211nnn/GSE211799/suppl/GSE211799%5Fadata%5Fatlas.h5ad.gz -O {RAW_DATA_DIR}/mpi.h5ad.gz\n",
    "!gunzip {RAW_DATA_DIR}/mpi.h5ad.gz -d {RAW_DATA_DIR}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "2f1ee67e-f5fd-476d-b95b-5f96c3279428",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 301796 × 31706\n",
       "    obs: 'study_sample', 'study', 'file', 'reference', 'size_factors_sample', 'phase_cyclone', 's_cyclone', 'g2m_cyclone', 'g1_cyclone', 'sex', 'ins_score', 'ins_high', 'gcg_score', 'gcg_high', 'sst_score', 'sst_high', 'ppy_score', 'ppy_high', 'cell_filtering', 'age', 'strain', 'tissue', 'technique', 'study_sample_design', 'cell_type', 'cell_type_multiplet', 'cell_subtype', 'cell_subtype_multiplet', 'design', 'size_factors_integrated', 'pre_cell_type_unified', 'pre_cell_type_original', 'study_parsed', 'cell_type_parsed', 'low_q', 'BETA-DATA_leiden_r1.5', 'BETA-DATA_leiden_r20', 'BETA-DATA_hc_gene_programs', 'BETA-DATA_hc_gene_programs_parsed', 'BETA-DATA_leiden_r1.5_parsed', 'BETA-DATA_leiden_r1.5_parsed_const', 'CXG-DATA_n_genes', 'CXG-DATA_mt_frac', 'CXG-DATA_doublet_score', 'CXG-DATA_log10_n_counts', 'CXG-DATA_age_approxDays', 'CXG-DATA_cell_subtype_immune_reannotatedIntegrated', 'CXG-DATA_cell_subtype_endothelial_reannotatedIntegrated', 'CXG-DATA_emptyDrops_LogProb_scaled', 'CXG-DATA_diabetes_model', 'CXG-DATA_chemical_stress', 'CXG-DATA_GEO_accession', 'CXG-DATA_sex_annotation', 'cell_type_integrated_v2', 'cell_type_integrated_v2_parsed'\n",
       "    var: 'gene_symbol', 'used_integration', 'gene_symbol_original_matched', 'CXG-DATA_feature_is_filtered', 'CXG-DATA_present_Fltp_2y', 'CXG-DATA_present_Fltp_adult', 'CXG-DATA_present_Fltp_P16', 'CXG-DATA_present_NOD', 'CXG-DATA_present_NOD_elimination', 'CXG-DATA_present_spikein_drug', 'CXG-DATA_present_embryo', 'CXG-DATA_present_VSG', 'CXG-DATA_present_STZ', 'gene_symbol_FINAL'\n",
       "    uns: 'BETA-DATA_hc_gene_programs_parsed_colors', 'BETA-DATA_hc_gene_programs_parsed_order', 'BETA-DATA_leiden_r1.5_parsed_colors', 'BETA-DATA_leiden_r1.5_parsed_order', 'cell_type_integrated_v2_colors', 'cell_type_integrated_v2_parsed_order', 'field_descriptions', 'study_colors', 'study_order', 'study_parsed_colors', 'study_parsed_order'\n",
       "    obsm: 'BETA-DATA_X_umap', 'BETA-DATA_X_umap_opt', 'X_integrated', 'X_umap'"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load raw data\n",
    "data = ad.read_h5ad(RAW_DATA_DIR / \"mpi.h5ad\")\n",
    "data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "ce63035b-24f0-4a73-b80d-9f3847735f4c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 301796 × 19888\n",
       "    obs: 'study_sample', 'study', 'file', 'reference', 'size_factors_sample', 'phase_cyclone', 's_cyclone', 'g2m_cyclone', 'g1_cyclone', 'sex', 'ins_score', 'ins_high', 'gcg_score', 'gcg_high', 'sst_score', 'sst_high', 'ppy_score', 'ppy_high', 'cell_filtering', 'age', 'strain', 'tissue', 'technique', 'study_sample_design', 'cell_type', 'cell_type_multiplet', 'cell_subtype', 'cell_subtype_multiplet', 'design', 'size_factors_integrated', 'pre_cell_type_unified', 'pre_cell_type_original', 'study_parsed', 'cell_type_parsed', 'low_q', 'BETA-DATA_leiden_r1.5', 'BETA-DATA_leiden_r20', 'BETA-DATA_hc_gene_programs', 'BETA-DATA_hc_gene_programs_parsed', 'BETA-DATA_leiden_r1.5_parsed', 'BETA-DATA_leiden_r1.5_parsed_const', 'CXG-DATA_n_genes', 'CXG-DATA_mt_frac', 'CXG-DATA_doublet_score', 'CXG-DATA_log10_n_counts', 'CXG-DATA_age_approxDays', 'CXG-DATA_cell_subtype_immune_reannotatedIntegrated', 'CXG-DATA_cell_subtype_endothelial_reannotatedIntegrated', 'CXG-DATA_emptyDrops_LogProb_scaled', 'CXG-DATA_diabetes_model', 'CXG-DATA_chemical_stress', 'CXG-DATA_GEO_accession', 'CXG-DATA_sex_annotation', 'cell_type_integrated_v2', 'cell_type_integrated_v2_parsed'\n",
       "    var: 'gene_symbol', 'used_integration', 'gene_symbol_original_matched', 'CXG-DATA_feature_is_filtered', 'CXG-DATA_present_Fltp_2y', 'CXG-DATA_present_Fltp_adult', 'CXG-DATA_present_Fltp_P16', 'CXG-DATA_present_NOD', 'CXG-DATA_present_NOD_elimination', 'CXG-DATA_present_spikein_drug', 'CXG-DATA_present_embryo', 'CXG-DATA_present_VSG', 'CXG-DATA_present_STZ', 'gene_symbol_FINAL', 'n_cells'\n",
       "    uns: 'BETA-DATA_hc_gene_programs_parsed_colors', 'BETA-DATA_hc_gene_programs_parsed_order', 'BETA-DATA_leiden_r1.5_parsed_colors', 'BETA-DATA_leiden_r1.5_parsed_order', 'cell_type_integrated_v2_colors', 'cell_type_integrated_v2_parsed_order', 'field_descriptions', 'study_colors', 'study_order', 'study_parsed_colors', 'study_parsed_order'\n",
       "    obsm: 'BETA-DATA_X_umap', 'BETA-DATA_X_umap_opt', 'X_integrated', 'X_umap'"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Filtering\n",
    "sc.pp.filter_genes(data, min_cells=100, inplace=True)\n",
    "data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "eb2b6ad2-0910-4d4c-b969-f7db13aa36be",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 301796 × 19888\n",
       "    obs: 'study_sample', 'study', 'file', 'reference', 'size_factors_sample', 'phase_cyclone', 's_cyclone', 'g2m_cyclone', 'g1_cyclone', 'sex', 'ins_score', 'ins_high', 'gcg_score', 'gcg_high', 'sst_score', 'sst_high', 'ppy_score', 'ppy_high', 'cell_filtering', 'age', 'strain', 'tissue', 'technique', 'study_sample_design', 'cell_type', 'cell_type_multiplet', 'cell_subtype', 'cell_subtype_multiplet', 'design', 'size_factors_integrated', 'pre_cell_type_unified', 'pre_cell_type_original', 'study_parsed', 'cell_type_parsed', 'low_q', 'BETA-DATA_leiden_r1.5', 'BETA-DATA_leiden_r20', 'BETA-DATA_hc_gene_programs', 'BETA-DATA_hc_gene_programs_parsed', 'BETA-DATA_leiden_r1.5_parsed', 'BETA-DATA_leiden_r1.5_parsed_const', 'CXG-DATA_n_genes', 'CXG-DATA_mt_frac', 'CXG-DATA_doublet_score', 'CXG-DATA_log10_n_counts', 'CXG-DATA_age_approxDays', 'CXG-DATA_cell_subtype_immune_reannotatedIntegrated', 'CXG-DATA_cell_subtype_endothelial_reannotatedIntegrated', 'CXG-DATA_emptyDrops_LogProb_scaled', 'CXG-DATA_diabetes_model', 'CXG-DATA_chemical_stress', 'CXG-DATA_GEO_accession', 'CXG-DATA_sex_annotation', 'celltype', 'cell_type_integrated_v2_parsed', 'species', 'task_celltype', 'batch'\n",
       "    var: 'gene_symbol', 'used_integration', 'gene_symbol_original_matched', 'CXG-DATA_feature_is_filtered', 'CXG-DATA_present_Fltp_2y', 'CXG-DATA_present_Fltp_adult', 'CXG-DATA_present_Fltp_P16', 'CXG-DATA_present_NOD', 'CXG-DATA_present_NOD_elimination', 'CXG-DATA_present_spikein_drug', 'CXG-DATA_present_embryo', 'CXG-DATA_present_VSG', 'CXG-DATA_present_STZ', 'gene_symbol_FINAL', 'n_cells', 'gene_ensembl'\n",
       "    uns: 'BETA-DATA_hc_gene_programs_parsed_colors', 'BETA-DATA_hc_gene_programs_parsed_order', 'BETA-DATA_leiden_r1.5_parsed_colors', 'BETA-DATA_leiden_r1.5_parsed_order', 'cell_type_integrated_v2_colors', 'cell_type_integrated_v2_parsed_order', 'field_descriptions', 'study_colors', 'study_order', 'study_parsed_colors', 'study_parsed_order', 'celltype_order', 'batch_order'\n",
       "    obsm: 'BETA-DATA_X_umap', 'BETA-DATA_X_umap_opt', 'X_integrated', 'X_umap'"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Standardize attributes\n",
    "data.obs[\"species\"] = \"Mus musculus\"\n",
    "data.obs.rename(columns={\"cell_type_integrated_v2\": \"celltype\"}, inplace=True)\n",
    "data.var.index.name = None\n",
    "data.var[\"gene_ensembl\"] = data.var.index\n",
    "\n",
    "le = LabelEncoder()\n",
    "\n",
    "data.obs[\"task_celltype\"] = le.fit_transform(data.obs[\"celltype\"])\n",
    "data.uns[\"celltype_order\"] = le.classes_.tolist()\n",
    "\n",
    "data.obs[\"batch\"] = le.fit_transform(data.obs[\"study_sample_design\"])\n",
    "data.uns[\"batch_order\"] = le.classes_.tolist()\n",
    "\n",
    "data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "8ae576dd-56a8-44b0-99c0-b08dc61c293b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save processed data\n",
    "data.write_h5ad(PROCESSED_DATA_DIR / \"mpi.h5ad\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
